{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script will take the DAVs that are flanking 100 bp of ISREs and fold the sequence around the DAV. \n",
    "\n",
    "Then it will use SNPFold and RNAsnp to find DAVs that are altering the local folding landscape. \n",
    "\n",
    "Specifically look for ones that altering the base pairing probability of ISREs themselves. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "numBasesToFold = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script bash\n",
    "# First we need to combine the files for HGMD Muts and ClinVar Muts and sort them\n",
    "cat ../temp/HGMD/HGMDMuts_Flanking_100bpISREs.txt ../temp/ClinVar/ClinVarMuts_Flanking_100bpISREs.txt > ../temp/Muts_Flanking_100bpISREs.txt\n",
    "sort -k6,6 -k7,7n -k8,8n ../temp/Muts_Flanking_100bpISREs.txt > ../temp/Muts_Flanking_100bpISREs_Sorted.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First we have to grab the coordinates +/- 100 base pairs of these mutations and write to file. \n",
    "muts = pd.read_table('../temp/Muts_Flanking_100bpISREs_Sorted.txt',sep=\"\\t\",header=None)\n",
    "print muts.head()\n",
    "muts_newdata = {\"chrm\":muts[5],\"start\":muts[6]-(numBasesToFold/2),\"end\":muts[7]+(numBasesToFold/2),\"snp_id\":muts[8],\"strand\":muts[9],\"score\":[0]*len(muts[8])}\n",
    "muts_newdf = pd.DataFrame(muts_newdata,columns=[\"chrm\",\"start\",\"end\",\"snp_id\",\"score\",\"strand\"]).drop_duplicates()\n",
    "print muts_newdf.head()\n",
    "muts_newdf.to_csv(\"../temp/Muts_Flanking_100bpISREs_\" + str(numBasesToFold) + \"bpwindow.txt\",header=False,index=False,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script bash\n",
    "numbases=100\n",
    "# Get the sequences for that file \n",
    "bedtools getfasta -fi /home/shared/hg38/hg38.fa -bed ../temp/Muts_Flanking_100bpISREs_${numbases}bpwindow.txt -name -s -tab -fo ../temp/Muts_Flanking_100bpISREs_${numbases}bpwindow-Sequences.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "# This function will run SNPfold on every mutation and write a file that contains the results with the mutation information\n",
    "# It requires a file that contains sequences around mutation, a file containing info of the WT base and MUT base\n",
    "def runSNPFold(mutseqfile, mutfile, baseOfinterest,writefile):\n",
    "    \n",
    "    # Read the file containing sequences around mutation\n",
    "    mutseqfile = pd.read_table(mutseqfile,sep=\"\\t\",header=None)\n",
    "    # Create a dictionary for those sequences using the mutation id as key\n",
    "    mutseq_dict = pd.Series(mutseqfile[1].values,index=mutseqfile[0].values).to_dict()\n",
    "    \n",
    "    # Read the file containing information about WT base and MUT base\n",
    "    mutfile = pd.read_table(mutfile,sep=\"\\t\",header=None).iloc[:,range(5,13)].drop_duplicates()\n",
    "    print mutfile.head()\n",
    "    # Create a dictionary for strand, WT base and MUT base using the mutation id as key\n",
    "    mut_dict = {row[8]:[row[9],row[10],row[11]] for index, row in mutfile.iterrows()} \n",
    "    #pd.Series(mutfile.iloc[:,[4,5,6]].values,index=mutfile[8].values).to_dict()\n",
    "\n",
    "    # Dictionary for complementary base dictionary\n",
    "    base_dict = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}\n",
    "    \n",
    "    # Go through every mutation and run SNPFold on it, the accurate version and write results to file\n",
    "    with open(writefile, \"w\") as fw:\n",
    "        # First write header\n",
    "        fw.write(\"MUT\" + \"\\t\" + \"CC_BPPROB\" + \"\\t\" + \"CC_BPPROB_PVAL\" + \"\\t\" + \"CC_SHANNON\" + \"\\t\" + \"CC_SHANNON_PVAL\")\n",
    "        fw.write(\"\\n\")\n",
    "        # Go through every mutation\n",
    "        for mut in mut_dict.keys():\n",
    "            # Get the sequence and bases for the mutation\n",
    "            seq4mut = mutseq_dict[mut]\n",
    "            bases4mut = mut_dict[mut]\n",
    "            # If the mutation is on the negative strand, get the complement of the WT and MUT base\n",
    "            # Else just get the WT and MUT base\n",
    "            if bases4mut[0]==\"-\":\n",
    "                WTbase=base_dict[bases4mut[1]]\n",
    "                MUTbase=base_dict[bases4mut[2]]\n",
    "            else:\n",
    "                WTbase=bases4mut[1]\n",
    "                MUTbase=bases4mut[2]\n",
    "            # Call SNP fold and write result into file    \n",
    "            pipe_out = subprocess.check_output([\"SNPfold_commandline.py\",\"-m\",\"all\",\"-n\",mut,\"-A\", seq4mut, WTbase + baseOfinterest + MUTbase])\n",
    "            snpfold_for_mut = [i.split('\\t') for i in pipe_out.strip().split('\\n')][1]\n",
    "            fw.write(mut + \"\\t\" + \"\\t\".join(snpfold_for_mut[1:]))\n",
    "            fw.write(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run function \n",
    "#runSNPFold(\"../temp/Muts_Flanking_100bpISREs_\"+str(numBasesToFold)+\"bpwindow-Sequences.txt\",\"../temp/Muts_Flanking_100bpISREs_Sorted.txt\",str((numBasesToFold/2)+1),\"../temp/Muts_Flanking_100bpISREs_\"+str(numBasesToFold)+\"bpwindow_SNPFoldResults.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(22268, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MUT</th>\n",
       "      <th>CC_BPPROB</th>\n",
       "      <th>CC_BPPROB_PVAL</th>\n",
       "      <th>CC_SHANNON</th>\n",
       "      <th>CC_SHANNON_PVAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>465460</td>\n",
       "      <td>0.678471</td>\n",
       "      <td>0.10526</td>\n",
       "      <td>0.690041</td>\n",
       "      <td>0.13816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>228057</td>\n",
       "      <td>0.995841</td>\n",
       "      <td>0.93421</td>\n",
       "      <td>0.970579</td>\n",
       "      <td>0.91118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>378465</td>\n",
       "      <td>0.998228</td>\n",
       "      <td>0.87500</td>\n",
       "      <td>0.977722</td>\n",
       "      <td>0.80921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>378462</td>\n",
       "      <td>0.862681</td>\n",
       "      <td>0.39474</td>\n",
       "      <td>0.825546</td>\n",
       "      <td>0.49013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>427285</td>\n",
       "      <td>0.952034</td>\n",
       "      <td>0.44737</td>\n",
       "      <td>0.899208</td>\n",
       "      <td>0.56908</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      MUT  CC_BPPROB  CC_BPPROB_PVAL  CC_SHANNON  CC_SHANNON_PVAL\n",
       "0  465460   0.678471         0.10526    0.690041          0.13816\n",
       "1  228057   0.995841         0.93421    0.970579          0.91118\n",
       "2  378465   0.998228         0.87500    0.977722          0.80921\n",
       "3  378462   0.862681         0.39474    0.825546          0.49013\n",
       "4  427285   0.952034         0.44737    0.899208          0.56908"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in SNPFold data\n",
    "SNPfold_data = pd.read_table(\"../temp/Muts_Flanking_100bpISREs_\"+str(numBasesToFold)+\"bpwindow_SNPFoldResults.txt\",sep=\"\\t\",header=0)\n",
    "print SNPfold_data.shape\n",
    "SNPfold_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 0.8]       5389\n",
      "(0.8, 0.9]     3902\n",
      "(0.9, 0.95]    3859\n",
      "(0.95, 1]      9111\n",
      "dtype: int64\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYcAAAEDCAYAAADeP8iwAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3XtcVXW+//HXFkQdFIiSzThhU2ZH\nxhQd70GHwoABdAQHbU5HSzpjjVlGmd18hJmXRmsmpxwbOdNldMY5pQU+klEU8kYjqWijls45OpnQ\nxKZRQCSVi9/fH/xcD52FuEU2G+j9fDx65F57r+/6fL8PWG/W7bsdxhiDiIjIBTp5uwAREWl7FA4i\nImKjcBARERuFg4iI2CgcRETERuEgIiI2vt4uoKUUFRV5uwQRkXZpyJAhtmUdJhyg8Q6KiMilXeoP\na51WEhERG4WDiIjYKBxERMRG4SAiIjYKBxERsVE4iIiIjcJBRERsFA4iImLToR6CExHZ+u/R3i6h\nzYjetrXZ6+rIQUREbBQOIiJio3AQEREbhYOIiNgoHERExEbhICIiNgoHERGxUTiIiIiNwkFERGwU\nDiIiYqNwEBERG4WDiIjYKBxERMRG4SAiIjYKBxERsVE4iIiIjcJBRERsFA4iImKjcBARERuFg4iI\n2CgcRETExqPh8Pbbb5OUlMSYMWN4/PHHOXv2LMXFxUyYMIHY2FjS09OpqakBoKamhvT0dGJjY5kw\nYQIlJSVWO8uXLyc2Npb4+Hi2b9/uyZJFRAQPhoPL5WLFihW89957rFu3jvr6enJycnj55ZeZMmUK\nmzZtIiAggDVr1gCwevVqAgIC2LRpE1OmTOHll18G4PDhw+Tk5JCTk8Pvfvc75s6dS319vafKFhER\nPHzkUF9fz5kzZ6irq+PMmTP07NmTwsJC4uPjAUhJSSE/Px+ADz/8kJSUFADi4+PZsWMHxhjy8/NJ\nSkrCz8+PsLAwbrjhBvbt2+fJskVEvvV8PdWw0+nk/vvv584776RLly5ERkbSv39/AgIC8PVt2Gxo\naCgulwtoONL47ne/21CUry89evSgvLwcl8tFRETERe2eX+dfHTx40FPdERFpd65mn+ixcKisrCQ/\nP5/8/Hx69OjBo48+6vHrBeHh4R5tX0TavjJvF9CGuLNPLCoqanS5x04r/eUvf+H6668nODiYzp07\nExcXx549ezh58iR1dXUAlJaW4nQ6gYYjgq+++gqAuro6qqqquOaaa3A6nZSWllrtulwuax0REfEM\nj4VDr169+Otf/8rp06cxxrBjxw5uvvlmRowYQW5uLgBZWVnExMQAEBMTQ1ZWFgC5ubmMHDkSh8NB\nTEwMOTk51NTUUFxczNGjRxk4cKCnyhYRETx4WikiIoL4+HhSUlLw9fUlPDycu+++mzvuuIPHHnuM\nJUuWEB4ezoQJEwBITU1l1qxZxMbGEhgYyCuvvAJA3759SUhIIDExER8fHzIyMvDx8fFU2SIiAjiM\nMcbbRbSEoqIihgwZ4u0yRMTLtv57tLdLaDOit2297Gcute/UE9IiImKjcBARERuFg4iI2CgcRETE\nRuEgIiI2CgcREbFROIiIiI3CQUREbBQOIiJio3AQEREbhYOIiNgoHERExEbhICIiNgoHERGxUTiI\niIiNwkFERGwUDiIiYqNwEBERG4WDiIjYKBxERMRG4SAiIjYKBxERsVE4iIiIjcJBRERsFA4iImKj\ncBARERuFg4iI2CgcRETERuEgIiI2CgcREbFROIiIiI3CQUREbBQOIiJio3AQEREbhYOIiNgoHERE\nxMaj4XDy5ElmzJjBj370IxISEti7dy8VFRWkpaURFxdHWloalZWVABhjmD9/PrGxsYwdO5ZPP/3U\naicrK4u4uDji4uLIysryZMkiIoKHw2HBggXcfvvtbNiwgbVr19KnTx8yMzMZNWoUGzduZNSoUWRm\nZgKwbds2jh49ysaNG5k3bx7PP/88ABUVFSxdupR3332X1atXs3TpUitQRETEMzwWDlVVVezatYvU\n1FQA/Pz8CAgIID8/n+TkZACSk5PJy8sDsJY7HA4GDRrEyZMnKSsro6CggMjISIKCgggMDCQyMpLt\n27d7qmwREQF8PdVwSUkJwcHBPPPMMxw6dIj+/fsze/Zsjh8/TkhICAA9e/bk+PHjALhcLkJDQ631\nQ0NDcblctuVOpxOXy+WpskVEBA+GQ11dHZ999hnPPfccERERzJ8/3zqFdJ7D4cDhcLTYNg8ePNhi\nbYmItHdXs0/0WDiEhoYSGhpKREQEAD/60Y/IzMzk2muvpaysjJCQEMrKyggODgYajghKS0ut9UtL\nS3E6nTidTnbu3Gktd7lcDB8+vNFthoeHe6o7ItJOlHm7gDbEnX1iUVFRo8s9ds2hZ8+ehIaG8ve/\n/x2AHTt20KdPH2JiYsjOzgYgOzub0aNHA1jLjTF88skn9OjRg5CQEKKioigoKKCyspLKykoKCgqI\nioryVNkiIoIHjxwAnnvuOZ544glqa2sJCwvjxRdf5Ny5c6Snp7NmzRp69erFkiVLAIiOjmbr1q3E\nxsbSrVs3Fi5cCEBQUBAPPfSQdWF7+vTpBAUFebJsEZFvPYcxxni7iJZQVFTEkCFDvF2GiHjZ1n+P\n9nYJbUb0tq2X/cyl9p16QlpERGwUDiIiYqNwEBERG7fCYf369Zw6dQqAZcuW8fDDD18095GIiHQs\nboXDsmXL6N69O7t372bHjh2kpqZacx+JiEjH41Y4+Pj4ALB161YmTpzIHXfcQW1trUcLExER73Er\nHJxOJxkZGfz5z38mOjqampoazp075+naRETES9wKhyVLlhAVFcUbb7xBQEAAFRUVPPnkk56uTURE\nvOSyT0jX19eTkpLChg0brGUhISHWzKoiItLxXDYcfHx8uPHGG/nHP/5Br169WqMmjxgya4W3S2gz\nil6619sliEgb59bcSidPniQpKYmBAwfSrVs3a/lvf/tbjxUmIiLe41Y4PProo56uQ0RE2hC3wmH4\n8OF8+eWXfPHFF9x2222cPn2a+vp6T9cmIiJe4tbdSu+++y4zZswgIyMDaPjCnenTp3u0MBER8R63\nwuGPf/wjf/rTn+jevTsA3//+9zlx4oRHCxMREe9xKxz8/Pzw8/OzXtfV1XmsIBER8T63rjkMGzaM\n3/72t5w5c4aPPvqIVatWERMT4+naRETES9w6cnjiiScIDg7mlltu4Z133iE6Opr09HRP1yYiIl7i\n1pFDp06dSE5OZuDAgTgcDm688UYcDoenaxMRES9xKxy2bNnCnDlz6N27N8YYSkpKmDt3LtHR+q5W\nEZGOyK1w+MUvfsGKFSu44YYbADh27BgPPPCAwkFEpINy65qDv7+/FQwAYWFh+Pv7e6woERHxriaP\nHDZu3AjArbfeytSpU0lISMDhcLBhwwYGDBjQKgWKiEjrazIcNm/ebP37uuuuY9euXQAEBwdz9uxZ\nz1YmIiJe02Q4vPjii61Vh4iItCFuXZAuLi7mD3/4A19++eVFT0drym4RkY7JrXCYPn06qamp3Hnn\nnXTq5NY1bBERacfcCocuXbpw77369jARkW8Lt8Lh3nvvZenSpURGRl40AV///v09VpiIiHiPW+Hw\nv//7v6xdu5bCwkJr2gyHw8GKFfpeZhGRjsitcNiwYQN5eXkXHTWIiEjH5dbV5b59+1JVVeXpWkRE\npI1w68ihqqqKhIQEBgwYQOfOna3lupVVRKRjciscHnnkEU/XISIibYhb4TB8+HBP1yEiIm2IW+Ew\nePBg6y6l2tpa6urq6NatG3v27PFocSIi4h1uhcPevXutfxtjyM/P55NPPvFYUSIi4l1XPBeGw+Hg\nrrvuoqCgwK3P19fXk5yczIMPPgg0zNM0YcIEYmNjSU9Pp6amBoCamhrS09OJjY1lwoQJlJSUWG0s\nX76c2NhY4uPj2b59+5WWLCIiV8itI4fz3+sAcO7cOQ4cOECXLl3c2sCKFSvo06cPp06dAuDll19m\nypQpJCUlkZGRwZo1a7jnnntYvXo1AQEBbNq0iZycHF5++WWWLFnC4cOHycnJIScnB5fLRVpaGrm5\nufj4+DSjuyIi4g63jhw2b95s/VdQUIC/vz/Lli277HqlpaVs2bKF1NRUoOGUVGFhIfHx8QCkpKSQ\nn58PwIcffkhKSgoA8fHx7NixwzqFlZSUhJ+fH2FhYdxwww3s27evWZ0VERH3uHXk0NzvdVi4cCGz\nZs2iuroagPLycgICAvD1bdhsaGgoLpcLAJfLxXe/+92Gonx96dGjB+Xl5bhcLiIiIqw2nU6ntc6/\nOnjwYLPq/LbROIl8O1zN73qT4bB06dJLvudwOJg+ffol39+8eTPBwcHceuutfPzxx80u8EqEh4c3\n8e6uVqmhPWh6nETatzJvF9CGuPO7XlRU1OjyJsPhO9/5jm3ZN998w3vvvUdFRUWT4bBnzx4+/PBD\ntm3bxtmzZzl16hQLFizg5MmT1NXV4evrS2lpKU6nE2g4Ivjqq68IDQ2lrq6OqqoqrrnmGpxOJ6Wl\npVa7LpfLWkdERDyjyXC4//77rX+fOnWKFStW8P7775OYmHjRe42ZOXMmM2fOBODjjz/mzTff5Je/\n/CUzZswgNzeXpKQksrKyiImJASAmJoasrCwGDx5Mbm4uI0eOxOFwEBMTw8yZM0lLS8PlcnH06FEG\nDhx4tf2Wq3TshQHeLqHN6J2x39sliLS4y15zqKio4K233uKDDz4gJSWFrKwsAgMDm73BWbNm8dhj\nj7FkyRLCw8OZMGECAKmpqcyaNYvY2FgCAwN55ZVXgIZJ/xISEkhMTMTHx4eMjAzdqSQdTuRrkd4u\noc346JGPvF2CcJlwWLRoEZs2bWLixIl88MEH+Pv7N2sjI0aMYMSIEQCEhYWxZs0a22e6dOnCq6++\n2uj606ZNY9q0ac3atoiIXLkmw+Gtt97Cz8+P119//aIZWI0xOBwOTZ8hItJBNRkOhw4daq06RESk\nDbni6TNERKTjUziIiIiNwkFERGwUDiIiYqNwEBERG4WDiIjYKBxERMRG4SAiIjYKBxERsVE4iIiI\njcJBRERsFA4iImKjcBARERuFg4iI2CgcRETERuEgIiI2CgcREbFROIiIiI3CQUREbBQOIiJio3AQ\nEREbhYOIiNgoHERExEbhICIiNgoHERGxUTiIiIiNwkFERGwUDiIiYqNwEBERG4WDiIjYKBxERMRG\n4SAiIjYKBxERsVE4iIiIjcfC4auvvmLy5MkkJiaSlJTE73//ewAqKipIS0sjLi6OtLQ0KisrATDG\nMH/+fGJjYxk7diyffvqp1VZWVhZxcXHExcWRlZXlqZJFROT/81g4+Pj48PTTT/PnP/+Zd955h1Wr\nVnH48GEyMzMZNWoUGzduZNSoUWRmZgKwbds2jh49ysaNG5k3bx7PP/880BAmS5cu5d1332X16tUs\nXbrUChQREfEMj4VDSEgI/fv3B6B79+7cdNNNuFwu8vPzSU5OBiA5OZm8vDwAa7nD4WDQoEGcPHmS\nsrIyCgoKiIyMJCgoiMDAQCIjI9m+fbunyhYREVrpmkNJSQkHDx4kIiKC48ePExISAkDPnj05fvw4\nAC6Xi9DQUGud0NBQXC6XbbnT6cTlcrVG2SIi31q+nt5AdXU1M2bM4Nlnn6V79+4XvedwOHA4HC22\nrYMHD7ZYWx1ZS4yTfwvU0VHo565laTxbztWMpUfDoba2lhkzZjB27Fji4uIAuPbaaykrKyMkJISy\nsjKCg4OBhiOC0tJSa93S0lKcTidOp5OdO3day10uF8OHD290e+Hh4U1Us+vqO9RBND1O7jnWAnV0\nFC0xnuRdfRMdxdWOZ1kL1dERuDOWRUVFjS732GklYwyzZ8/mpptuIi0tzVoeExNDdnY2ANnZ2Ywe\nPfqi5cYYPvnkE3r06EFISAhRUVEUFBRQWVlJZWUlBQUFREVFeapsERHBg0cORUVFrF27lltuuYVx\n48YB8Pjjj/PAAw+Qnp7OmjVr6NWrF0uWLAEgOjqarVu3EhsbS7du3Vi4cCEAQUFBPPTQQ6SmpgIw\nffp0goKCPFW2iIjgwXAYOnQof/vb3xp97/wzDxdyOBzMmTOn0c+npqZa4SAiIp6nJ6RFRMRG4SAi\nIjYKBxERsVE4iIiIjcJBRERsFA4iImKjcBARERuFg4iI2CgcRETERuEgIiI2CgcREbFROIiIiI3C\nQUREbBQOIiJio3AQEREbhYOIiNgoHERExEbhICIiNgoHERGxUTiIiIiNwkFERGwUDiIiYqNwEBER\nG4WDiIjYKBxERMRG4SAiIjYKBxERsVE4iIiIjcJBRERsFA4iImKjcBARERuFg4iI2CgcRETERuEg\nIiI2CgcREbFROIiIiI3CQUREbNpNOGzbto34+HhiY2PJzMz0djkiIh1auwiH+vp6XnjhBX73u9+R\nk5PDunXrOHz4sLfLEhHpsNpFOOzbt48bbriBsLAw/Pz8SEpKIj8/39tliYh0WA5jjPF2EZezYcMG\ntm/fzoIFCwDIzs5m3759ZGRkWJ8pKiryVnkiIu3akCFDbMt8vVCHRzTWORERaZ52cVrJ6XRSWlpq\nvXa5XDidTi9WJCLSsbWLcBgwYABHjx6luLiYmpoacnJyiImJ8XZZIiIdVrs4reTr60tGRgY/+9nP\nqK+v5yc/+Ql9+/b1dlkiIh1WuzhyAIiOjiY3N5e8vDymTZvmlRrOnDnDpEmTqK+vByArK4u4uDji\n4uLIysq67PoVFRWkpaURFxdHWloalZWVjX5u8eLFJCUlkZCQwPz58zl/z8DkyZMZPHgw+/fvb7lO\ntZLmjN3BgweZOHEi48aNY/z48ezbt++y23Gn3UOHDnH33XczduxYfv7zn3Pq1CkAdu/eTWJiImPG\njGlmLz2nOeN3qX42xZ3nib788kvuu+8+xo4dy+TJky865RseHs64ceMYN24cP//5z63lM2fOZPjw\n4WzYsOFKut1qWnJ8S0pKGDhwoDUOF944cynr168nKSmJfv36XfT77dWfSSNu+8Mf/mDefvttY4wx\n5eXlJiYmxpSXl5uKigoTExNjKioqmlx/0aJFZvny5cYYY5YvX24WL15s+0xRUZG5++67TV1dnamr\nqzMTJ040hYWF1vuTJk0y+/bta8FetY7mjF1aWprZsmWLMcaYLVu2mEmTJjW5DXfbHT9+vPn444+N\nMcasXr3avPLKK9Z7xcXFJikpqdn99JTmjF9T/WxMXV2dGT16tDl27Jg5e/asGTt2rPm///s/2+ce\neeQR8/777xtjjPnLX/5innjiCeu9QYMGXbL9p556yqxfv/7ynfWClhzf5vwMHT582Bw5cqTR329v\n/Uy2myOHtuCDDz5g9OjRABQUFBAZGUlQUBCBgYFERkayffv2JtfPz88nOTkZgOTkZPLy8myfcTgc\n1NTUUFtba/3/uuuua/nOtLLmjJ3D4aC6uhqAqqoqQkJCmtyGu+0ePXqUYcOGARAZGcnGjRuvtnse\n15zxu9J+uvs80ZEjRxg5ciQAI0eO7BDPHLXG+DalT58+3HTTTc1e3xMUDm6qqamhuLiY66+/Hmi4\nYyo0NNR63+l04nK5mmzj+PHj1g6uZ8+eHD9+3PaZwYMHM2LECKKiooiKiuL222+nT58+LdiT1tfc\nsXv22WdZvHgx0dHRLFq0iMcff7zJ7bjbbt++fa0d2oYNG/jqq6+a1a/W0tzxu9J+uttuv379rB3h\npk2bqK6upry8HICzZ88yfvx4Jk6c2OgfP22RJ8a3pKSE5ORkJk2axO7duz3cA89QOLipvLycHj16\ntFh7DocDh8NhW/7FF19w5MgRtm7dyrZt2ygsLGy3P1znNXfs/vSnP/HMM8+wdetWnnnmGWbPnt0i\n9SxYsIBVq1Yxfvx4qqur8fPza5F2PaW54+epfj755JPs2rWL5ORkdu7cidPpxMfHB4DNmzfz/vvv\n88tf/pKFCxdy7NixFtmmJ7X0+IaEhLB582ays7N5+umnmTlzplvXe9oahYObunbtSk1NjfW6Oc9e\nXHvttZSVlQFQVlZGcHCw7TObNm0iIiICf39//P39uf3229m7d28L9cI7mjt25y8KAiQkJFz2grS7\n7fbp04c333yT999/n6SkJMLCwq64T62pueN3pf10t12n08nSpUvJzs7mscceAyAgIMB6DyAsLIzh\nw4fz2WefXUFPvaOlx9fPz49rrrkGgFtvvZXevXvz+eefe7gXLU/h4KbAwEDq6+s5e/YsAFFRURQU\nFFBZWUllZSUFBQVERUUBDX9ZNbYji4mJITs7G2iYAuT8Oc4L9erVi127dlFXV0dtbS27du1q96eV\nrmTsLhQSEsLOnTsBKCws5Pvf/z7QcG78ySeftH3e3XbPn847d+4cr7/+Oj/96U9bqqse0dzxu1Q/\nXS4X9913n+3z7j5PdOLECc6dOwdAZmYmP/nJTwCorKy0drInTpxgz5493HzzzS0wAp7V0uN74sQJ\n666n4uJijh49agXHpfYNbVG7eM6hrYiMjKSoqIjbbruNoKAgHnroIVJTUwGYPn06QUFBAPztb39r\n9OLpAw88QHp6OmvWrKFXr14sWbIEgP379/M///M/LFiwgPj4eAoLCxk7diwOh4Pbb7+9Qzzw5+7Y\nzZ49m5/+9KcMGDCAefPmsXDhQurq6ujSpQsvvPACAP/4xz/o2rWrbRvutrtu3TpWrVoFQGxsrLVz\na8uaM36X6mdZWRm+vvZf/aaeJ/r1r3/NrbfeyujRo9m5cye/+tWvcDgcDB06lDlz5gANF6rnzJmD\nw+HAGMPUqVPbRThAy47vrl27ePXVV/H19aVTp07MnTv3svuGTZs2MW/ePE6cOMGDDz5IeHg4b7zx\nRmt0/dJa/f6oduzAgQMX3bbXmKqqKvPII494rIb2eiurO2Pnrl/84hfm4MGDLdLWv2qrt7K25Pit\nXLnS5OXltUhbV6It38rakuN7Kc3dN+hW1nagf//+jBgxwjpkbEz37t159dVXPbL9yZMnU1xcTOfO\nnT3Svie5M3bueuqpp+jXr18LVHWx3bt3M23aNOt8cVvSkuM3adKkRk9petLMmTPZuXMnXbp0adXt\nuqslx/dSmrNv8ObPZLuYsltERFqXjhxERMRG4SAiIjYKB2kV5ydk+/GPf0xKSgp79uwBLp6kLDEx\nkYyMDOs2yQt9/fXXPPbYY9x1112MHz+eqVOneuze8cmTJ192csO3336b06dPW6+nTp3KyZMnW2T7\n2dnZjBkzhrFjx5KcnNzsu1ZOnDjBhAkTSE5OZvfu3axfv56EhASrf/Pnz29y/avpU15enr7nvb1r\n9Uvg8q104YRs27ZtM//5n/9pjLn4Toza2lpzzz33mNzc3IvWPXfunJk4caJZtWqVtezgwYNm165d\nbm27trbW1l59ff0lP+/OHWF33nmnOX78uFvbvxJbtmwxycnJprS01BhjzNmzZ80777zTrLbWrVtn\nnn32Wev1/fff7/aYXa22fGeSuEfPOUirO3XqlPVE7YV8fX0ZPHgwX3zxxUXLCwsL8fX15T/+4z+s\nZefvVjLGsHjxYrZv347D4WDatGkkJiby8ccf8+tf/5qAgAA+//xz3njjDf7rv/6LiIgIPv30UzIz\nM/n888957bXXqKmpISwsjBdffBF/f/+Ltj1nzhz279/P2bNniY+PZ8aMGaxYsYKysjLuu+8+goKC\nWLlyJTExMaxZs4bg4GDeeust3nvvPQBSU1OZMmUKJSUlTJ06lSFDhrB3716cTifLli2zPa+RmZnJ\nk08+aT2R6+fnx8SJE4GGKcznzJnD6dOn6d27NwsXLiQwMJBjx44xd+5cysvL6dq1K/PmzaOmpoaX\nXnqJM2fOcODAAWJjY9mzZw+zZ88mJiaGO+64gzfffJPly5dTXV3N/PnzOXDgAAAPP/ww8fHxF/Vp\n7dq1rFy5ktraWiIiIpgzZw4+Pj4MHjyYe++9l82bN9O1a1eWLVvGsWPH+PDDD9m5cyevv/46r732\nGr17976aHxnxBm+nk3w79OvXz/z4xz828fHx5oc//KHZv3+/MebiI4dvvvnGjB8/3pqm+7zf//73\nZsGCBY22u2HDBjNlyhRTV1dnvv76axMdHW1cLpcpLCw0ERER5tixY9Z2/u3f/s3s3bvXGGPM8ePH\nzT333GOqq6uNMQ1TqL/22mvGmIuPHMrLy40xDdNZT5o0yXq+4l+PHM6/3r9/vxkzZoyprq42p06d\nMomJiebTTz81xcXFJjw83Hz22WfGGGNmzJhhsrOzbf0ZNmyYOXnyZKN9HTNmjDVF9JIlS8z8+fON\nMcbce++95vPPPzfGGPPJJ5+YyZMnG2OMee+998zcuXOt9S/sV2FhoXnggQeMMcYsXrzYassYY01P\nfb5Phw8fNg8++KCpqakxxhgzZ84ck5WVZYwx5pZbbjH5+fnGmIYp6X/zm98YY3Tk0BHoyEFaRdeu\nXVm7di0Ae/fu5amnnmLdunUAHDt2jHHjxuFwOBg9ejTR0dFut1tUVERSUhI+Pj5cd911DBs2jP37\n99O9e3cGDBhw0XxCvXr1YtCgQQD89a9/5fDhw9bRSG1trfXehdavX8+7775LXV0dX3/9NUeOHGny\nGYuioiLuuusuvvOd7wANT87u3r2bmJgYrr/+esLDw4GG++q//PJLt/tZVVVFVVUVw4cPByAlJYVH\nH32U6upq9u7dy6OPPmp99sJ5gtyxY8cOfvWrX1mvAwMDbe8fOHDAemL4zJkzXHvttQB07tyZO++8\nE2iYR+ijjz66om1L26VwkFY3ePBgysvLOXHiBAC9e/e2gqMxffv2JTc394q3c34H3dhrYwyRkZEX\n7RT/VXFxMW+++SZr1qwhMDCQp59+2pp/pzkunBXVx8en0bZuvvlmDhw4wKhRo9xq0xhDQEBAk+N3\ntYwxpKSkMHPmTNt7nTt3tmYX7tSpk0cfIpPWpbuVpNUdOXKE+vp6a76Zyxk5ciQ1NTW888471rJD\nhw6xe/duhg4dyvr166mvr+fEiRPs3r2bgQMHXrbNQYMGsWfPHuv6xjfffGO7+6m6uppu3brRo0cP\n/vnPf7Jt2zbrPX9/f+uLiC4pVp37AAABwElEQVQ0dOhQ8vLyOH36NN988w15eXkMHTrUrX4CPPjg\ng7z00kt8/fXXQMNRwOrVq+nRowcBAQHW9O1r165l2LBhdO/eneuvv57169cDDTvyQ4cOub09gNtu\nu40//vGP1ut//fraUaNGkZuba000V1FRcdmjnkuNj7QfOnKQVnHmzBnGjRsHNOzAFi1aZH0HwOU4\nHA6WLl3KwoUL+e///m+6dOnC9773PZ599lnrAu/501KzZs2iZ8+e/P3vf2+yzeDgYF588UUef/xx\n6zRMeno6N954o/WZfv368YMf/ICEhARCQ0P54Q9/aL03ceJEfvaznxESEsLKlSut5f3792f8+PFM\nmDABaLgg/YMf/ICSkhK3+hodHc0///lP0tLSMMbgcDisCd0WLVpkXZA+fwEd4KWXXuL555/n9ddf\np66ujsTExCuaXmTatGm88MILjBkzhk6dOvHwww9bU6VDw9FMeno6999/P+fOnaNz585kZGTwve99\n75JtJiYm8txzz7Fy5UpeffVVXZBuhzR9hoiI2Oi0koiI2CgcRETERuEgIiI2CgcREbFROIiIiI3C\nQUREbBQOIiJio3AQERGb/wcow2/0EEqfzQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fc9964ab610>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot correlation coefficient data as bar plot\n",
    "corr_data = SNPfold_data[\"CC_BPPROB\"]\n",
    "corr_data_breakdown = pd.cut(np.array(corr_data),bins=[0,0.8,0.9,0.95,1])\n",
    "corr_data_breakdown_freq = corr_data_breakdown.value_counts()\n",
    "print corr_data_breakdown_freq\n",
    "counts_df = pd.DataFrame({\"Numbers\":corr_data_breakdown_freq.values,\"BP Correlation Coefficient\":corr_data_breakdown_freq.index.values},columns=[\"BP Correlation Coefficient\",\"Numbers\"])\n",
    "sns.set_style(\"whitegrid\")\n",
    "ax = sns.barplot(x=\"BP Correlation Coefficient\",y=\"Numbers\",data=counts_df)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5396, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MUT</th>\n",
       "      <th>CC_BPPROB</th>\n",
       "      <th>CC_BPPROB_PVAL</th>\n",
       "      <th>CC_SHANNON</th>\n",
       "      <th>CC_SHANNON_PVAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>465460</td>\n",
       "      <td>0.678471</td>\n",
       "      <td>0.10526</td>\n",
       "      <td>0.690041</td>\n",
       "      <td>0.13816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>378460</td>\n",
       "      <td>0.673778</td>\n",
       "      <td>0.16776</td>\n",
       "      <td>0.870998</td>\n",
       "      <td>0.57237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>287149</td>\n",
       "      <td>0.685392</td>\n",
       "      <td>0.16776</td>\n",
       "      <td>0.356607</td>\n",
       "      <td>0.08224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>82448</td>\n",
       "      <td>0.787902</td>\n",
       "      <td>0.53289</td>\n",
       "      <td>0.822089</td>\n",
       "      <td>0.54276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CM095436</td>\n",
       "      <td>0.770476</td>\n",
       "      <td>0.09539</td>\n",
       "      <td>0.658545</td>\n",
       "      <td>0.11513</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         MUT  CC_BPPROB  CC_BPPROB_PVAL  CC_SHANNON  CC_SHANNON_PVAL\n",
       "0     465460   0.678471         0.10526    0.690041          0.13816\n",
       "5     378460   0.673778         0.16776    0.870998          0.57237\n",
       "7     287149   0.685392         0.16776    0.356607          0.08224\n",
       "10     82448   0.787902         0.53289    0.822089          0.54276\n",
       "11  CM095436   0.770476         0.09539    0.658545          0.11513"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grab the mutations that result in correlation less than 0.8\n",
    "SNPfold_data_lowCC = SNPfold_data[SNPfold_data[\"CC_BPPROB\"]<=0.8]\n",
    "print SNPfold_data_lowCC.shape\n",
    "SNPfold_data_lowCC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We want a function that will go through the list of mutations that result in low BP CC and grab ones that change\n",
    "# the BP of ISREs nearby \n",
    "def getMutsAlterISRE_BP(ISREmutfile,SNPfoldfile,cc_cutoff,foldingdist):\n",
    "    print foldingdist\n",
    "    # These are the lists that will contain the muts that have an ISRE nearby that changes BP probability or NOT\n",
    "    muts_StrucISRE_changes =[]\n",
    "    muts_StrucISRE_DontChange =[]\n",
    "    # Dictionary for complementary base dictionary\n",
    "    base_dict = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}\n",
    "    # Read the ISREs near Muts file\n",
    "    ISREs_nearMuts = pd.read_table(ISREmutfile,sep=\"\\t\",header=None)\n",
    "    # Read the SNPFold results file\n",
    "    SNPfold_results = pd.read_table(SNPfoldfile,sep=\"\\t\",header=0)\n",
    "    # Get the muts that results in CC of BP to be less than the cut-off\n",
    "    SNPfold_lowCC = SNPfold_results[SNPfold_results[\"CC_BPPROB\"]<=cc_cutoff]\n",
    "    # Go through every mutation\n",
    "    for mut in SNPfold_lowCC[\"MUT\"].values:\n",
    "        # Get all the ISREs for the Mut\n",
    "        ISREs_for_Mut = ISREs_nearMuts[ISREs_nearMuts[8]==mut]\n",
    "        # Get the base probability file for the mut\n",
    "        #bp_for_Mut = pd.read_table(\"../scripts/output/\"+mut+'_'+str(foldingdist+1)+\"/input_mutations_bpProbs_perNt.txt\",sep=\"\\t\",header=None)\n",
    "        bp_for_Mut = pd.read_table(\"../scripts/output/\"+mut+\"/input_mutations_bpProbs_perNt.txt\",sep=\"\\t\",header=None)\n",
    "        # Get the mutation for the MUT id, this so that we can get the right row from bpProbabilities table\n",
    "        if ISREs_for_Mut.iloc[0,9]==\"-\":\n",
    "            WTbase=base_dict[ISREs_for_Mut.iloc[0,10]]\n",
    "            MUTbase=base_dict[ISREs_for_Mut.iloc[0,11]]\n",
    "        else:\n",
    "            WTbase=ISREs_for_Mut.iloc[0,10]\n",
    "            MUTbase=ISREs_for_Mut.iloc[0,11]\n",
    "        if WTbase==\"T\":\n",
    "            WTbase=\"U\"\n",
    "        if MUTbase==\"T\":\n",
    "            MUTbase=\"U\"\n",
    "        # Get only the WT row and the mutant row\n",
    "        WT_bp = bp_for_Mut[bp_for_Mut[0]==\"WT\"]\n",
    "        MUT_bp = bp_for_Mut[bp_for_Mut[0]==(WTbase+str(foldingdist+1)+MUTbase)]\n",
    "        # This keeps count of how many ISREs that have changes in bp probs \n",
    "        numISREs_Mut = 0\n",
    "        # Go through every ISRE for Mut and get the base pairing probabilities for WT seq and MUT seq for ISRE\n",
    "        # If the location of ISRE is less than 6 bases away from the start and end of the seq\n",
    "        # then reduce the number of bases we look at for that ISRE\n",
    "        for index,row in ISREs_for_Mut.iterrows():\n",
    "            ISRE = row[3]\n",
    "            # If minus strand, the ISRE coordinates are flipped, the start coordinate is actually the end coordinate\n",
    "            # And if ISRE coordinate is greater than mut, it is actually on the left of the Mut instead of the right \n",
    "            # of the mut\n",
    "            if row[9] == \"-\":\n",
    "                ISRE_dist_from_Mut = row[1]+50+5 - row[6]\n",
    "                start_index_for_ISRE = foldingdist+1 - ISRE_dist_from_Mut\n",
    "            else:\n",
    "                ISRE_dist_from_Mut = row[1]+50 - row[6]\n",
    "                start_index_for_ISRE = foldingdist+1 + ISRE_dist_from_Mut\n",
    "            if start_index_for_ISRE > (2*foldingdist)-4:\n",
    "                bpProb_WT = WT_bp.iloc[0,range(start_index_for_ISRE,(2*foldingdist)+2)]\n",
    "                bpProb_MUT = MUT_bp.iloc[0,range(start_index_for_ISRE,(2*foldingdist)+2)]\n",
    "            elif start_index_for_ISRE < 6:\n",
    "                bpProb_WT = WT_bp.iloc[0,range(1,start_index_for_ISRE+1)]\n",
    "                bpProb_MUT = MUT_bp.iloc[0,range(1,start_index_for_ISRE+1)]\n",
    "            else:\n",
    "                bpProb_WT = WT_bp.iloc[0,range(start_index_for_ISRE,start_index_for_ISRE+6)]\n",
    "                bpProb_MUT = MUT_bp.iloc[0,range(start_index_for_ISRE,start_index_for_ISRE+6)]\n",
    "            # Number of bases for ISRE whose bp probability differs greater than 0.5 between WT and MUT seq\n",
    "            Diffs = [i for i in bpProb_WT-bpProb_MUT if abs(i)>=0.5]\n",
    "            if len(Diffs)>0:\n",
    "                muts_StrucISRE_changes.append([mut,ISRE,ISRE_dist_from_Mut,len(Diffs)])\n",
    "                numISREs_Mut += 1\n",
    "        # Check if no ISREs have bp probability changes between WT and MUT seq, then add MUT to dont change list\n",
    "        if numISREs_Mut == 0 :\n",
    "            muts_StrucISRE_DontChange.append(mut)\n",
    "    return muts_StrucISRE_changes,muts_StrucISRE_DontChange"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50\n"
     ]
    }
   ],
   "source": [
    "# Use a correlation coefficient of 1 to look for all mutations regardless of how they change local structure\n",
    "# We just want to grab any Muts that are altering ISRE bp Probabilities \n",
    "muts_altering_ISREbpProb,muts_NOTaltering_ISREbpProb = getMutsAlterISRE_BP(\"../temp/Muts_Flanking_100bpISREs_Sorted.txt\",\"../temp/Muts_Flanking_100bpISREs_\"+str(numBasesToFold)+\"bpwindow_SNPFoldResults.txt\",1,numBasesToFold/2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of DAVs altering ISRE bp Probability:  4688\n",
      "Number of DAVs NOT altering ISRE bp Probability:  17579\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY4AAAEDCAYAAAAiKuN6AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAIABJREFUeJzt3X9YlHW+//HnCJGEBoIyHDlgulGH\nY5gt2sZK0E6i8hsVdd3N0tVc87e27cZ6Fg1L7ep0DppXJtFVunk6FaEEo0lQiT+21NlcpWg9liSa\nDHuhgOYigvP9o6+zsv7iVoZBfT2uy6uZz/3r/UG6X34+9z33mBwOhwMREZE26uLuAkRE5Pqi4BAR\nEUMUHCIiYoiCQ0REDFFwiIiIIQoOERExxNPdBbiazWZzdwkiItelyMjIi7bf8MEBl+68iIhc3OX+\n0a2pKhERMUTBISIihig4RETEEAWHiIgYouAQERFDXHZXVUZGBp988gkBAQEUFRUBMHfuXA4ePAjA\niRMn6N69OwUFBRw+fJiEhAT69u0LwL333ktWVhYA5eXlZGRk0NjYSGxsLAsWLMBkMlFXV8e8efM4\ncuQIwcHBZGdn4+vr66ruiIjI/+eyEceoUaPIzc1t1ZadnU1BQQEFBQUMGzaMuLg457LQ0FDnsnOh\nAbBo0SIWL15McXExlZWVlJWVAZCTk0NUVBTFxcVERUWRk5Pjqq6IiMh5XBYcgwcPvuQIwOFwsGnT\nJpKSki67j5qaGk6ePMnAgQMxmUykpaVRWloKQGlpKWlpaQCkpaVRUlLSvh0QEZGLcssHAHfv3k1A\nQAB33HGHs+3w4cOkpaXRrVs35s6dy6BBg7Db7QQFBTnXCQoKwm63A1BbW0tgYCAAvXr1ora2tkP7\nINIZHMqKcHcJ0gmFZu5z6f7dEhxFRUWtRhuBgYF8/PHH9OjRg/LycmbMmIHVam3z/kwmEyaT6ZLL\nKyoqrqlekc7Kx90FSKfk6nNehwdHc3MzH374Ifn5+c42Ly8vvLy8ALjnnnsIDQ3l4MGDmM1mqqur\nnetVV1djNpsBCAgIoKamhsDAQGpqavD397/kMcPDw13UGxH3OuTuAqRTao9zXqd65MiOHTvo169f\nqymoY8eO0dLSAkBVVRWVlZWEhIQQGBhIt27d2LNnDw6Hgw0bNvDwww8DYLFY2LBhA0CrdhERcS2X\njTjmz5/Pzp07OX78ODExMcyaNYsxY8awceNGEhMTW627a9cuVqxYgaenJ126dOGZZ57Bz88PgIUL\nFzpvx42JiSEmJgaAqVOnMnfuXPLy8ujduzfZ2dmu6oqIiJzH5HA4HO4uwpVsNpuejis3LF0cl4tp\nj4vjlzt36pPjIiJiiIJDREQMUXCIiIghCg4RETFEwSEiIoYoOERExBAFh4iIGKLgEBERQxQcIiJi\niIJDREQMUXCIiIghCg4RETFEwSEiIoYoOERExBAFh4iIGKLgEBERQxQcIiJiiIJDREQMUXCIiIgh\nCg4RETFEwSEiIoa4LDgyMjKIiooiKSnJ2fbSSy/x4IMPkpqaSmpqKlu2bHEuW716NXFxcQwfPpyt\nW7c628vKyhg+fDhxcXHk5OQ426uqqhgzZgxxcXHMnTuXpqYmV3VFRETO47LgGDVqFLm5uRe0T5w4\nkYKCAgoKCoiNjQXgwIEDWK1WrFYrubm5PPPMM7S0tNDS0kJWVha5ublYrVaKioo4cOAAAP/5n//J\nxIkT+fDDD7n99tvJy8tzVVdEROQ8LguOwYMH4+vr26Z1S0tLSUxMxMvLi5CQEPr06cPevXvZu3cv\nffr0ISQkBC8vLxITEyktLcXhcPDpp58yfPhwAEaOHElpaamruiIiIufp8Gsc69atIzk5mYyMDOrr\n6wGw2+0EBQU51zGbzdjt9ku2Hz9+nNtvvx1PT08AgoKCsNvtHdsREZGblGdHHmz8+PFMnz4dk8nE\n8uXLWbZsGUuXLnX5cSsqKlx+DBF38HF3AdIpufqc16HB0bNnT+frMWPGMG3aNOCHkUR1dbVzmd1u\nx2w2A1y0vUePHjQ0NNDc3IynpyfV1dXO9S8mPDy8vbsi0ikccncB0im1xznPZrNdclmHTlXV1NQ4\nX5eUlBAWFgaAxWLBarXS1NREVVUVlZWVDBgwgIiICCorK6mqqqKpqQmr1YrFYsFkMvGTn/yEzZs3\nA7B+/XosFktHdkVE5KblshHH/Pnz2blzJ8ePHycmJoZZs2axc+dOvvrqKwCCg4PJysoCICwsjPj4\neBISEvDw8CAzMxMPDw8AMjMzmTJlCi0tLYwePdoZNk899RTz5s0jOzub8PBwxowZ46quiIjIeUwO\nh8Ph7iJcyWazERkZ6e4yRFziUFaEu0uQTig0c9817+Ny5059clxERAxRcIiIiCEKDhERMUTBISIi\nhig4RETEEAWHiIgYouAQERFDFBwiImKIgkNERAxRcIiIiCEKDhERMUTBISIihig4RETEEAWHiIgY\nouAQERFDFBwiImKIgkNERAxRcIiIiCEKDhERMUTBISIihrgsODIyMoiKiiIpKcnZ9vzzzzNixAiS\nk5OZMWMGDQ0NABw+fJgBAwaQmppKamoqmZmZzm3Ky8tJTk4mLi6OZ599FofDAUBdXR2TJk1i2LBh\nTJo0ifr6eld1RUREzuOy4Bg1ahS5ubmt2oYMGUJRURGFhYXccccdrF692rksNDSUgoICCgoKyMrK\ncrYvWrSIxYsXU1xcTGVlJWVlZQDk5OQQFRVFcXExUVFR5OTkuKorIiJyHpcFx+DBg/H19W3VFh0d\njaenJwADBw6kurr6svuoqanh5MmTDBw4EJPJRFpaGqWlpQCUlpaSlpYGQFpaGiUlJS7ohYiI/DO3\nXeN47733iImJcb4/fPgwaWlpPPLII+zevRsAu91OUFCQc52goCDsdjsAtbW1BAYGAtCrVy9qa2s7\nsHoRkZuXpzsOumrVKjw8PEhJSQEgMDCQjz/+mB49elBeXs6MGTOwWq1t3p/JZMJkMl1yeUVFxTXX\nLNIZ+bi7AOmUXH3O6/DgyM/P55NPPuGNN95wnuy9vLzw8vIC4J577iE0NJSDBw9iNptbTWdVV1dj\nNpsBCAgIoKamhsDAQGpqavD397/kMcPDw13YIxH3OeTuAqRTao9zns1mu+SyDp2qKisrIzc3l1Wr\nVuHt7e1sP3bsGC0tLQBUVVVRWVlJSEgIgYGBdOvWjT179uBwONiwYQMPP/wwABaLhQ0bNgC0ahcR\nEddy2Yhj/vz57Ny5k+PHjxMTE8OsWbPIycmhqamJSZMmAXDvvfeSlZXFrl27WLFiBZ6ennTp0oVn\nnnkGPz8/ABYuXEhGRgaNjY3ExMQ4r4tMnTqVuXPnkpeXR+/evcnOznZVV0RE5Dwmx7kPRtygbDYb\nkZGR7i5DxCUOZUW4uwTphEIz913zPi537tQnx0VExBAFh4iIGKLgEBERQxQcIiJiiIJDREQMUXCI\niIghCg4RETFEwSEiIoYoOERExBAFh4iIGKLgEBERQxQcIiJiiIJDREQMUXCIiIghbQqOTZs2cfLk\nSQBefvllZs6cyRdffOHSwkREpHNqU3C8/PLLdOvWjd27d/OnP/2J9PR0Fi1a5OLSRESkM2pTcHh4\neACwZcsWxo4dy0MPPcSZM2dcWpiIiHRObQoOs9lMZmYmGzduJDY2lqamJs6ePevq2kREpBNqU3Bk\nZ2cTHR3Na6+9xu23305dXR2//e1vXV2biIh0Qp5XWqGlpYWRI0fywQcfONsCAwMJDAx0aWEiItI5\nXXHE4eHhQd++ffnuu+86oh4REenkrjjiAGhoaCAxMZEBAwbg7e3tbH/llVcuu11GRgaffPIJAQEB\nFBUVAVBXV8e8efM4cuQIwcHBZGdn4+vri8Ph4LnnnmPLli107dqVZcuW0b9/fwDWr1/PqlWrAHji\niScYOXIkAOXl5WRkZNDY2EhsbCwLFizAZDIZ/ymIiEibtSk45syZc1U7HzVqFI888gi/+93vnG05\nOTlERUUxdepUcnJyyMnJ4amnnqKsrIzKykqKi4v5y1/+wqJFi3j33Xepq6tj5cqVvPfee5hMJkaN\nGoXFYsHX15dFixaxePFi7r33Xh5//HHKysqIjY29qlpFRKRt2nRx/P777yc4OJjm5mbuv/9+IiIi\n+Pd///crbjd48GB8fX1btZWWlpKWlgZAWloaJSUlrdpNJhMDBw6koaGBmpoatm3bxpAhQ/Dz88PX\n15chQ4awdetWampqOHnyJAMHDsRkMpGWlkZpaanR/ouIiEFtGnG88847vP3229TX11NSUoLdbmfh\nwoWsWbPG8AFra2udF9Z79epFbW0tAHa7naCgIOd6QUFB2O32C9rNZvNF28+tfzEVFRWG6xS5Hvi4\nuwDplFx9zmtTcKxbt453332XsWPHAnDHHXdw7Nixaz64yWTqkGsS4eHhLj+GiDsccncB0im1xznP\nZrNdclmbpqq8vLzw8vJyvm9ubr7qYgICAqipqQGgpqYGf39/4IeRRHV1tXO96upqzGbzBe12u/2i\n7efWFxER12pTcAwePJhXXnmFxsZGtm/fzpw5c7BYLFd1QIvFwoYNGwDYsGEDDz/8cKt2h8PBnj17\n6N69O4GBgURHR7Nt2zbq6+upr69n27ZtREdHExgYSLdu3dizZw8Oh6PVvkRExHXaNFX1m9/8hry8\nPO666y7efvttYmNjGTNmzBW3mz9/Pjt37uT48ePExMQwa9Yspk6dyty5c8nLy6N3795kZ2cDEBsb\ny5YtW4iLi8Pb25slS5YA4Ofnx/Tp00lPTwdgxowZ+Pn5AbBw4ULn7bgxMTHExMRc1Q9BRETazuRw\nOBxtWbGpqYlvvvkGk8lE3759W01ddWY2m43IyEh3lyHiEoeyItxdgnRCoZn7rnkflzt3tmnE8ckn\nn7Bw4UJCQ0NxOBwcPnyYZ555Rp+ZEBG5CbUpOJYtW8batWvp06cPAIcOHWLq1KkKDhGRm1CbLo77\n+Pg4QwMgJCQEHx/dQS4icjO67IijuLgYgHvuuYfHH3+c+Ph4TCYTH3zwARERmlsVEbkZXTY4Pv74\nY+frnj17smvXLgD8/f05ffq0aysTEZFO6bLBsXTp0o6qQ0RErhNtujheVVXFm2++yZEjR1p9avxK\nj1UXEZEbT5uCY8aMGaSnp/Ozn/2MLl3adD1dRERuUG0KjltvvZVHH33U1bWIiMh1oE3B8eijj7Jy\n5UqGDBnS6hPj576hT0REbh5tCo79+/dTUFDAp59+6nwMuslkYu3atS4tTkREOp82BccHH3xASUnJ\ndfN8KhERcZ02XekOCwvjxIkTrq5FRESuA20acZw4cYL4+HgiIiK45ZZbnO26HVdE5ObTpuCYNWuW\nq+sQEZHrRJuC4/7773d1HSIicp1oU3Dcd999zrupzpw5Q3NzM97e3vz5z392aXEiItL5tCk4Pv/8\nc+drh8NBaWkpe/bscVlRIiLSeRl+fojJZGLo0KFs27bNFfWIiEgn16YRx7nv5QA4e/Ys5eXl3Hrr\nrS4rSkREOq82Bcf538vh4eFBcHAwL7/88lUd8JtvvmHevHnO91VVVcyePZsTJ07wzjvv4O/vD8D8\n+fOdX027evVq8vLy6NKlC//xH//Bgw8+CEBZWRnPPfccZ8+eZcyYMUydOvWqahIRkbZrU3C05/dy\n9OvXj4KCAgBaWlqIiYkhLi6O/Px8Jk6cyOTJk1utf+DAAaxWK1arFbvdzqRJk9i8eTMAWVlZvP76\n65jNZtLT07FYLNx5553tVquIiFzossGxcuXKSy4zmUzMmDHjmg7+pz/9iZCQEIKDgy+5TmlpKYmJ\niXh5eRESEkKfPn3Yu3cvAH369CEkJASAxMRESktLFRwiIi522Yvjt9122wV/AN577z1yc3Ov+eBW\nq5WkpCTn+3Xr1pGcnExGRgb19fUA2O12goKCnOuYzWbsdvsl20VExLUuO+L41a9+5Xx98uRJ1q5d\nS35+PgkJCa2WXY2mpiY++ugjnnzySQDGjx/P9OnTMZlMLF++nGXLlrXbFFlFRUW77Eeks/FxdwHS\nKbn6nHfFaxx1dXW8/vrrFBYWMnLkSNavX4+vr+81H7isrIz+/fvTs2dPAOd/AcaMGcO0adOAH0YS\n1dXVzmV2ux2z2QxwyfZ/Fh4efs31inRGh9xdgHRK7XHOs9lsl1x22amq559/nvT0dHx8fCgsLGTW\nrFntEhrwwzRVYmKi831NTY3zdUlJCWFhYQBYLBasVitNTU1UVVVRWVnJgAEDiIiIoLKykqqqKpqa\nmrBarVgslnapTURELu2yI47XX38dLy8vVq1a1epJuA6HA5PJdNWPHDl16hQ7duwgKyvL2fbCCy/w\n1VdfARAcHOxcFhYWRnx8PAkJCXh4eJCZmYmHhwcAmZmZTJkyhZaWFkaPHu0MGxERcR2Tw+FwuLsI\nV7LZbERGRrq7DBGXOJQV4e4SpBMKzdx3zfu43LnT8CNHRETk5qbgEBERQxQcIiJiiIJDREQMUXCI\niIghCg4RETFEwSEiIoYoOERExBAFh4iIGKLgEBERQxQcIiJiiIJDREQMUXCIiIghCg4RETFEwSEi\nIoYoOERExBAFh4iIGKLgEBERQxQcIiJiiIJDREQMUXCIiIghnu46sMViwcfHhy5duuDh4UF+fj51\ndXXMmzePI0eOEBwcTHZ2Nr6+vjgcDp577jm2bNlC165dWbZsGf379wdg/fr1rFq1CoAnnniCkSNH\nuqtLIiI3BbeOONasWUNBQQH5+fkA5OTkEBUVRXFxMVFRUeTk5ABQVlZGZWUlxcXFLF68mEWLFgFQ\nV1fHypUreeedd3j33XdZuXIl9fX17uqOiMhNoVNNVZWWlpKWlgZAWloaJSUlrdpNJhMDBw6koaGB\nmpoatm3bxpAhQ/Dz88PX15chQ4awdetWd3ZBROSG57apKoDJkydjMpkYN24c48aNo7a2lsDAQAB6\n9epFbW0tAHa7naCgIOd2QUFB2O32C9rNZjN2u/2C41RUVLi4JyLu4ePuAqRTcvU5z23B8dZbb2E2\nm6mtrWXSpEn069ev1XKTyYTJZGqXY4WHh7fLfkQ6m0PuLkA6pfY459lstksuc9tUldlsBiAgIIC4\nuDj27t1LQEAANTU1ANTU1ODv7+9ct7q62rltdXU1ZrP5gna73e7cr4iIuIZbguPUqVOcPHnS+Xr7\n9u2EhYVhsVjYsGEDABs2bODhhx8GcLY7HA727NlD9+7dCQwMJDo6mm3btlFfX099fT3btm0jOjra\nHV0SEblpuGWqqra2lhkzZgDQ0tJCUlISMTExREREMHfuXPLy8ujduzfZ2dkAxMbGsmXLFuLi4vD2\n9mbJkiUA+Pn5MX36dNLT0wGYMWMGfn5+7V5v5FNr232fcv2zvfCou0sQcQuTw+FwuLsIV7LZbERG\nRl7TPhQccjGdITgOZUW4uwTphEIz913zPi537uxUt+OKiEjnp+AQERFDFBwiImKIgkNERAxRcIiI\niCEKDhERMUTBISIihig4RETEEAWHiIgYouAQERFDFBwiImKIgkNERAxRcIiIiCEKDhERMUTBISIi\nhig4RETEEAWHiIgYouAQERFDFBwiImKIgkNERAzp8OA4evQoEyZMICEhgcTERNasWQPASy+9xIMP\nPkhqaiqpqals2bLFuc3q1auJi4tj+PDhbN261dleVlbG8OHDiYuLIycnp6O7IiJyU/Ls6AN6eHjw\n9NNP079/f06ePMno0aMZMmQIABMnTmTy5Mmt1j9w4ABWqxWr1YrdbmfSpEls3rwZgKysLF5//XXM\nZjPp6elYLBbuvPPOju6SiMhNpcODIzAwkMDAQAC6detGv379sNvtl1y/tLSUxMREvLy8CAkJoU+f\nPuzduxeAPn36EBISAkBiYiKlpaUKDhERF3PrNY7Dhw9TUVHBvffeC8C6detITk4mIyOD+vp6AOx2\nO0FBQc5tzGYzdrv9ku0iIuJaHT7iOOf7779n9uzZ/P73v6dbt26MHz+e6dOnYzKZWL58OcuWLWPp\n0qXtcqyKiop22Y/I+TrD75WPuwuQTsnVv5tuCY4zZ84we/ZskpOTGTZsGAA9e/Z0Lh8zZgzTpk0D\nfhhJVFdXO5fZ7XbMZjPAJdv/WXh4+DVWvOsat5cb0bX/Xl27Q+4uQDql9vjdtNlsl1zW4VNVDoeD\nBQsW0K9fPyZNmuRsr6mpcb4uKSkhLCwMAIvFgtVqpampiaqqKiorKxkwYAARERFUVlZSVVVFU1MT\nVqsVi8XS0d0REbnpdPiIw2azUVBQwF133UVqaioA8+fPp6ioiK+++gqA4OBgsrKyAAgLCyM+Pp6E\nhAQ8PDzIzMzEw8MDgMzMTKZMmUJLSwujR492ho2IiLhOhwfHoEGD+Otf/3pBe2xs7CW3eeKJJ3ji\niScuus3lthMRkfanT46LiIghCg4RETFEwSEiIoYoOERExBAFh4iIGKLgEBERQxQcIiJiiIJDREQM\nUXCIiIghCg4RETFEwSEiIoYoOERExBAFh4iIGKLgEBERQxQcIiJiiIJDREQMUXCIiIghCg4RETFE\nwSEiIoYoOERExBAFh4iIGHLdB0dZWRnDhw8nLi6OnJwcd5cjInLDu66Do6WlhaysLHJzc7FarRQV\nFXHgwAF3lyUickO7roNj79699OnTh5CQELy8vEhMTKS0tNTdZYmI3NA83V3AtbDb7QQFBTnfm81m\n9u7de8F6Npvtmo6T8/P+17S93Jiu9feqXSS+4e4KpBP6m4t/N6/r4GiLyMhId5cgInJDua6nqsxm\nM9XV1c73drsds9nsxopERG5813VwREREUFlZSVVVFU1NTVitViwWi7vLEhG5oV3XweHp6UlmZiZT\npkwhISGB+Ph4wsLC3F3WdcvhcDB+/Hi2bNnibNu0aROTJ092Y1UiF7r77rtZtmyZ8/1rr73GSy+9\n5MaKbi4mh8PhcHcR0nns37+fOXPmsGHDBpqbmxk5ciS5ubmEhoa6uzQRp4iICHr16kVeXh7+/v68\n9tprnDp1ilmzZrm7tJvCDX9xXIy56667+NnPfsarr77KqVOnSE1NJTQ0lPXr17Nu3TrOnDnDfffd\nR2ZmJmfPniUjI4OvvvoKh8PB2LFjefTRR93dBbkJeHp6Mm7cONasWcO8efNaLTt8+DC///3vOX78\nOP7+/ixdupTevXu7qdIb03U9VSWuMXPmTAoLC9m6dSuPP/44+/fv58MPP+R///d/KSgooKWlBavV\nyhdffMHx48cpLCykqKiItLQ0d5cuN5Ff/vKXFBYWcuLEiVbtzz77LCNHjqSwsJDk5GSeffZZN1V4\n49KIQy5w2223kZCQwG233YaXlxc7duxg3759jB49GoDGxkaCgoKIjo7m4MGDPPvss8TGxhIdHe3m\nyuVm0q1bN1JTU1m7di1du3Z1tn/++efO6x2pqam88MIL7irxhqXgkIvq0qULXbr8Y0A6evRo5s6d\ne8F677//PmVlZaxbt47i4mIWL17ckWXKTe6xxx5j1KhRjBo1yt2l3FQ0VSVXFBUVxaZNmzh27BgA\nx48f57vvvuPYsWM4HA7i4+OZM2cOX3zxhZsrlZuNn58fI0aMIC8vz9l23333YbVaASgsLGTQoEHu\nKu+GpRGHXNHdd9/NzJkzmTRpEmfPnuWWW25h0aJFeHh4sGDBAhwOByaTid/85jfuLlVuQr/61a9Y\nt26d8/0f/vAHMjIyeO2115wXx6V96XZcERExRFNVIiJiiIJDREQMUXCIiIghCg4RETFEwSEiIoYo\nOKTDlZSUcPfdd/P111872w4fPkxSUhIAFRUVrZ7Q21Z2u53Zs2e3S435+flkZWUB8M033zBhwgRS\nU1OJj4/nD3/4AwCfffYZkZGRpKamMmLECJ5//vlW2z/wwAOkpqY6/xw4cKDVMc7v89XWeO4YCQkJ\nvPPOO1fdx7a67777Ltq+fPlyduzYAcCECRPYt28fAI8//jgNDQ00NDS0umVWrm/6HId0uKKiIiIj\nI7FarRc90VdUVFBeXk5sbGyb99nc3IzZbGbFihXtWSoAzz33HI899hhDhw4F4K9//atz2aBBg1i9\nejWNjY2kpaUxdOhQ57dOJiQkkJmZ2e71nO/cMWpra0lMTMRisdCzZ0/n8ubmZjw9Xf+/+Zw5cy7a\n/uqrrwI/hORbb73FL3/5S5fXIq6n4JAO9f3332Oz2Vi7di3Tpk27IDiamppYsWIFjY2N2Gw2fv3r\nX/PQQw+xePFi/u///o/m5mZmzpzJ0KFDyc/Pp7i4mFOnTnH27FmWLVvGtGnTKCoqIj8/n48++oi/\n//3vVFVVMXToUH77298C8O6775Kbm0v37t35t3/7N7y8vC57gq+pqWn13fZ33333Bet07dqV8PBw\n7Ha7oZ9Hc3MzTz75JF9++SVhYWE8//zzeHt7Y7FYGDFiBFu3buXWW2/lxRdfpE+fPpfcT0BAAKGh\noXz33Xe89dZbHDp0iKqqKnr37s3SpUtZtGgR5eXleHh48PTTT/PAAw8AcPToUSZMmIDdbiclJYWZ\nM2cCMH36dKqrqzl9+jSPPvoo48aNcx5ryZIlbN++nZ49e/Lf//3f+Pv78/TTT/PQQw8xYsSIVnVZ\nLBby8vJ48cUXOXToEKmpqfz0pz+ltraWYcOGOcP4ySefJD4+3vleOjdNVUmHKi0t5cEHH6Rv3770\n6NGD8vLyVsu9vLyYPXs2CQkJFBQUkJCQwCuvvMIDDzxAXl4ea9eu5YUXXuDUqVMAfPnll6xYsYI3\n33zzgmNVVFSQnZ1NYWEhmzZt4ujRo9jtdlatWsXbb7/NW2+9xTfffHPFmidOnMhjjz3GlClTeOON\nN2hoaLhgnfr6er799lsGDx7sbNu4cWOrqarGxsYLtjt48CC/+MUv2LRpEz4+PvzP//yPc1n37t0p\nLCzkkUceYcmSJZetsaqqiqqqKuf3pnz99de88cYb/Nd//ZdziqiwsJAXX3yRp59+mtOnTwOwb98+\nVqxYwfvvv88HH3zgnGJasmQJ+fn5vPfee/zxj3/k+PHjAJw6dYp77rkHq9XK4MGDWbly5RV/fvBD\nMISGhlJQUMDvfvc70tPTyc/MUzBdAAAELElEQVTPB+DEiRN8/vnnPPTQQ23al7ifgkM6lNVqJTEx\nEfhhmuXcM4UuZ9u2bbz66qukpqYyYcIETp8+zdGjRwEYMmQIfn5+F90uKiqK7t27c+utt/KjH/2I\nI0eOsG/fPgYPHoyfnx+33HLLBf9CvpjRo0ezceNGRowYwWeffcbYsWNpamoCYPfu3aSkpBATE0N0\ndDS9evVybncu/M79Of8Jruf8y7/8i3NqKyUlBZvN5lx27vpHYmIie/bsuWht58Jp/vz5ZGVlOX8W\nFovFeTybzUZKSgoAP/rRj+jduzcHDx4E4Kc//Sk9evSga9euxMXFOY//xz/+kZSUFMaOHcvRo0f5\n9ttvgR8efpmQkAD88OTZ8+s14v777+fbb7/l2LFjFBUVMXz48A6ZUpP2ob8p6TB1dXV8+umn7N+/\nH5PJREtLCyaTyTmFdDkrVqygX79+rdr+8pe/4O3tfcltvLy8nK89PDxoaWm56trNZjPp6emkp6eT\nlJTE/v37gX9c46iqqmLcuHHEx8cTHh7e5v2aTKbLvr+SS11HudzP5UrH/+yzz9ixYwdvv/023t7e\nzrBuy/ZGpKam8v7772O1WvU8qeuMRhzSYTZv3kxqaioff/wxH330EVu2bOFf//Vf2b17d6v1fHx8\n+P77753vo6OjefPNNzn3WLUvv/zyqmuIiIhg165d1NfX09zcTHFx8RW3KSsr48yZMwD87W9/o66u\nDrPZ3GqdkJAQpk6d6rwY3Fbfffcdn3/+OfCPmwbO2bRpE/DDqOJSdzO1xaBBgygsLAR+mBo7evSo\nM4S3b99OXV0djY2NlJSU8OMf/5gTJ07g6+uLt7c3X3/9davRztmzZ9m8eTPww9TX+fVezj//nQKM\nGjWKNWvWAHDnnXdedf+k4yk4pMMUFRVdcPFz2LBhFBUVtWr7yU9+woEDB0hNTWXjxo1Mnz6d5uZm\nUlJSSExMZPny5Vddg9ls5te//jVjxoxh/PjxBAcH071798tus337dpKSkkhJSWHy5Mk89dRTraak\nzvn5z3/Orl27OHz4MHDhNY4///nPF2zTt29f1q1bR3x8PA0NDYwfP965rL6+nuTkZNauXUtGRsZV\n9/kXv/gFDoeD5ORk5s2bx9KlS52jsQEDBjBr1ixSUlIYPnw4ERERxMTE0NzcTHx8PC+++CIDBw50\n7uu2225j7969JCUl8emnnzJjxow21dCjRw9+/OMfk5SU5LxtuWfPnvTr10/fpXEd0tNx5abz/fff\n4+Pj47xDa/To0cTFxbm7rFbO3Y3k7+/v7lJc5u9//zvJycmsX7/+iuEtnYuucchNZ+XKlezYsYPT\np08THR2tW0DdYMeOHSxYsIDHHntMoXEd0ohDREQM0TUOERExRMEhIiKGKDhERMQQBYeIiBii4BAR\nEUMUHCIiYsj/A6uVoLmoJf1CAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fc99a0b4c10>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Get the numbers of MUTs altering base pairing probability and plot as bar plot\n",
    "actualMuts_altering_ISREbpProb = list(set([i[0] for i in muts_altering_ISREbpProb]))\n",
    "print \"Number of DAVs altering ISRE bp Probability: \", len(actualMuts_altering_ISREbpProb)\n",
    "print \"Number of DAVs NOT altering ISRE bp Probability: \", len(muts_NOTaltering_ISREbpProb)\n",
    "counts_df = pd.DataFrame({\"Numbers\":[len(actualMuts_altering_ISREbpProb),len(muts_NOTaltering_ISREbpProb)],\"Altering ISRE bp Probability\":[\"Yes\",\"No\"]},columns=[\"Altering ISRE bp Probability\",\"Numbers\"])\n",
    "sns.set_style(\"whitegrid\")\n",
    "ax = sns.barplot(x=\"Altering ISRE bp Probability\",y=\"Numbers\",data=counts_df)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the mutations that are altering ISRE base pairing probability into a file\n",
    "pd.DataFrame(muts_altering_ISREbpProb).to_csv(\"../temp/Muts_Flanking_100bpISREs_\"+str(numBasesToFold)+\"bpwindow_AlteringISREbpProb.txt\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of ClinVar muts altering ISRE probability:  8125\n",
      "Number of HGMD muts altering ISRE probability:  10180\n"
     ]
    }
   ],
   "source": [
    "# Seperate out by ClinVar and HGMD mutations\n",
    "ClinVar_muts_altering_ISREbpProb = [i for i in muts_altering_ISREbpProb if i[0].isdigit()]\n",
    "print \"Number of ClinVar muts altering ISRE probability: \", len(ClinVar_muts_altering_ISREbpProb)\n",
    "HGMD_muts_altering_ISREbpProb = [i for i in muts_altering_ISREbpProb if ~i[0].isdigit()]\n",
    "print \"Number of HGMD muts altering ISRE probability: \", len(HGMD_muts_altering_ISREbpProb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1008, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>1</td>\n",
       "      <td>8957144</td>\n",
       "      <td>8957145</td>\n",
       "      <td>CM065045</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DP;MUT=ALT;GENE=CA6;STRAND=+;DNA=NM_0012...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>903</th>\n",
       "      <td>1</td>\n",
       "      <td>15505999</td>\n",
       "      <td>15506000</td>\n",
       "      <td>CM156160</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM?;MUT=ALT;GENE=CASP9;STRAND=-;DNA=NM_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1455</th>\n",
       "      <td>1</td>\n",
       "      <td>21561137</td>\n",
       "      <td>21561138</td>\n",
       "      <td>CM980058</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM;MUT=ALT;GENE=ALPL;STRAND=+;DNA=NM_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2123</th>\n",
       "      <td>1</td>\n",
       "      <td>25306615</td>\n",
       "      <td>25306616</td>\n",
       "      <td>CS158952</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=FP;MUT=ALT;GENE=RHD;STRAND=+;DNA=NM_0161...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2124</th>\n",
       "      <td>1</td>\n",
       "      <td>25306623</td>\n",
       "      <td>25306624</td>\n",
       "      <td>CM158949</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=FP;MUT=ALT;GENE=RHD;STRAND=+;DNA=NM_0161...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0         1         2         3  4  5  6  7  \\\n",
       "299   1   8957144   8957145  CM065045  .  A  G  .   \n",
       "903   1  15505999  15506000  CM156160  .  T  G  .   \n",
       "1455  1  21561137  21561138  CM980058  .  G  A  .   \n",
       "2123  1  25306615  25306616  CS158952  .  G  A  .   \n",
       "2124  1  25306623  25306624  CM158949  .  C  T  .   \n",
       "\n",
       "                                                      8  \n",
       "299   CLASS=DP;MUT=ALT;GENE=CA6;STRAND=+;DNA=NM_0012...  \n",
       "903   CLASS=DM?;MUT=ALT;GENE=CASP9;STRAND=-;DNA=NM_0...  \n",
       "1455  CLASS=DM;MUT=ALT;GENE=ALPL;STRAND=+;DNA=NM_000...  \n",
       "2123  CLASS=FP;MUT=ALT;GENE=RHD;STRAND=+;DNA=NM_0161...  \n",
       "2124  CLASS=FP;MUT=ALT;GENE=RHD;STRAND=+;DNA=NM_0161...  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get details about the HGMD mutations\n",
    "all_hgmd_muts = pd.read_csv(\"../processed_data/HGMD_PRO_2016.3_hg38.bed\", sep=\"\\t\",header=None,low_memory=False)\n",
    "hgmd_muts_relevent = all_hgmd_muts[all_hgmd_muts[3].isin([i[0] for i in HGMD_muts_altering_ISREbpProb])]\n",
    "print hgmd_muts_relevent.shape\n",
    "hgmd_muts_relevent.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2055, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2_x</th>\n",
       "      <th>3_x</th>\n",
       "      <th>0_y</th>\n",
       "      <th>1_y</th>\n",
       "      <th>2_y</th>\n",
       "      <th>3_y</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CM095436</td>\n",
       "      <td>CM095436</td>\n",
       "      <td>AGGGTT</td>\n",
       "      <td>-26</td>\n",
       "      <td>3</td>\n",
       "      <td>chr5</td>\n",
       "      <td>70076544</td>\n",
       "      <td>70076545</td>\n",
       "      <td>CM095436</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM;MUT=ALT;GENE=SMN2;STRAND=+;DNA=NM_017...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CM095436</td>\n",
       "      <td>CM095436</td>\n",
       "      <td>GGGTTT</td>\n",
       "      <td>-25</td>\n",
       "      <td>3</td>\n",
       "      <td>chr5</td>\n",
       "      <td>70076544</td>\n",
       "      <td>70076545</td>\n",
       "      <td>CM095436</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM;MUT=ALT;GENE=SMN2;STRAND=+;DNA=NM_017...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CM095436</td>\n",
       "      <td>CM095436</td>\n",
       "      <td>GGTTTT</td>\n",
       "      <td>-24</td>\n",
       "      <td>2</td>\n",
       "      <td>chr5</td>\n",
       "      <td>70076544</td>\n",
       "      <td>70076545</td>\n",
       "      <td>CM095436</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM;MUT=ALT;GENE=SMN2;STRAND=+;DNA=NM_017...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CM095436</td>\n",
       "      <td>CM095436</td>\n",
       "      <td>TTTAGA</td>\n",
       "      <td>-21</td>\n",
       "      <td>4</td>\n",
       "      <td>chr5</td>\n",
       "      <td>70076544</td>\n",
       "      <td>70076545</td>\n",
       "      <td>CM095436</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM;MUT=ALT;GENE=SMN2;STRAND=+;DNA=NM_017...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CM134391</td>\n",
       "      <td>CM134391</td>\n",
       "      <td>AGGAAC</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2877253</td>\n",
       "      <td>2877254</td>\n",
       "      <td>CM134391</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM?;MUT=ALT;GENE=GYG2;STRAND=+;DNA=NM_00...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          0       0_x     1_x  2_x  3_x   0_y       1_y       2_y       3_y  \\\n",
       "0  CM095436  CM095436  AGGGTT  -26    3  chr5  70076544  70076545  CM095436   \n",
       "1  CM095436  CM095436  GGGTTT  -25    3  chr5  70076544  70076545  CM095436   \n",
       "2  CM095436  CM095436  GGTTTT  -24    2  chr5  70076544  70076545  CM095436   \n",
       "3  CM095436  CM095436  TTTAGA  -21    4  chr5  70076544  70076545  CM095436   \n",
       "4  CM134391  CM134391  AGGAAC    1    1  chrX   2877253   2877254  CM134391   \n",
       "\n",
       "   4  5  6  7                                                  8  \n",
       "0  .  G  C  .  CLASS=DM;MUT=ALT;GENE=SMN2;STRAND=+;DNA=NM_017...  \n",
       "1  .  G  C  .  CLASS=DM;MUT=ALT;GENE=SMN2;STRAND=+;DNA=NM_017...  \n",
       "2  .  G  C  .  CLASS=DM;MUT=ALT;GENE=SMN2;STRAND=+;DNA=NM_017...  \n",
       "3  .  G  C  .  CLASS=DM;MUT=ALT;GENE=SMN2;STRAND=+;DNA=NM_017...  \n",
       "4  .  C  T  .  CLASS=DM?;MUT=ALT;GENE=GYG2;STRAND=+;DNA=NM_00...  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We want to merge the HGMD data together: the info about the mutation and how mutation is affecting ISREs\n",
    "HGMD_muts_altering_ISREbpProb_DF = pd.DataFrame(HGMD_muts_altering_ISREbpProb)\n",
    "HGMD_muts_altering_ISREbpProb_DF.head()\n",
    "combined_HGMD_data = HGMD_muts_altering_ISREbpProb_DF.merge(hgmd_muts_relevent,left_on=0,right_on=3)\n",
    "combined_HGMD_data['0_y'] = 'chr' + combined_HGMD_data['0_y'].astype(str)\n",
    "print combined_HGMD_data.shape\n",
    "combined_HGMD_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write HGMD data to a file\n",
    "combined_HGMD_data.iloc[:,range(1,14)].to_csv(\"../processed_data/Putative_RiboSPLitches_HGMDMuts_\"+str(numBasesToFold)+\"bpwindow.txt\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3680, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>1</td>\n",
       "      <td>1050053</td>\n",
       "      <td>1050054</td>\n",
       "      <td>263190</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.00051;AF_TGP=0.0016;ALLELEID=249340;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td>1</td>\n",
       "      <td>2303416</td>\n",
       "      <td>2303417</td>\n",
       "      <td>258899</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.00001;ALLELEID=249763;CLNDISDB=MedGe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>615</th>\n",
       "      <td>1</td>\n",
       "      <td>2559360</td>\n",
       "      <td>2559361</td>\n",
       "      <td>133401</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=137140;CLNDISDB=MedGen:CN169374;CLNDN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>620</th>\n",
       "      <td>1</td>\n",
       "      <td>2559502</td>\n",
       "      <td>2559503</td>\n",
       "      <td>133406</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=137145;CLNDISDB=MedGen:CN169374;CLNDN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>1</td>\n",
       "      <td>2559724</td>\n",
       "      <td>2559725</td>\n",
       "      <td>133413</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.01977;AF_TGP=0.0116;ALLELEID=137152;...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     0        1        2       3  4  5  6  7  \\\n",
       "154  1  1050053  1050054  263190  .  T  A  .   \n",
       "445  1  2303416  2303417  258899  .  C  T  .   \n",
       "615  1  2559360  2559361  133401  .  G  T  .   \n",
       "620  1  2559502  2559503  133406  .  C  A  .   \n",
       "627  1  2559724  2559725  133413  .  G  A  .   \n",
       "\n",
       "                                                     8  \n",
       "154  AF_EXAC=0.00051;AF_TGP=0.0016;ALLELEID=249340;...  \n",
       "445  AF_EXAC=0.00001;ALLELEID=249763;CLNDISDB=MedGe...  \n",
       "615  ALLELEID=137140;CLNDISDB=MedGen:CN169374;CLNDN...  \n",
       "620  ALLELEID=137145;CLNDISDB=MedGen:CN169374;CLNDN...  \n",
       "627  AF_EXAC=0.01977;AF_TGP=0.0116;ALLELEID=137152;...  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get details about the ClinVar mutations\n",
    "all_clinvar_muts = pd.read_csv(\"../processed_data/clinvar.bed\",sep=\"\\t\",header=None,low_memory=False)\n",
    "# Get the clinvar mutations\n",
    "clinvar_muts_relevent = all_clinvar_muts[all_clinvar_muts[3].isin([int(i[0]) for i in ClinVar_muts_altering_ISREbpProb])]\n",
    "print clinvar_muts_relevent.shape\n",
    "clinvar_muts_relevent.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8125, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2_x</th>\n",
       "      <th>3_x</th>\n",
       "      <th>0_y</th>\n",
       "      <th>1_y</th>\n",
       "      <th>2_y</th>\n",
       "      <th>3_y</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>228050</td>\n",
       "      <td>228050</td>\n",
       "      <td>AGGTAA</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>chr21</td>\n",
       "      <td>44504766</td>\n",
       "      <td>44504767</td>\n",
       "      <td>228050</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.00003;ALLELEID=231377;CLNDISDB=MedGe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>260298</td>\n",
       "      <td>260298</td>\n",
       "      <td>GGGAGG</td>\n",
       "      <td>-12</td>\n",
       "      <td>1</td>\n",
       "      <td>chr14</td>\n",
       "      <td>77279773</td>\n",
       "      <td>77279774</td>\n",
       "      <td>260298</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.0440;AF_EXAC=0.05999;AF_TGP=0.0184;AL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>195517</td>\n",
       "      <td>195517</td>\n",
       "      <td>AGGTAA</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10533501</td>\n",
       "      <td>10533502</td>\n",
       "      <td>195517</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.0002;AF_EXAC=0.00093;ALLELEID=192678;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>392112</td>\n",
       "      <td>392112</td>\n",
       "      <td>GAGTGT</td>\n",
       "      <td>-14</td>\n",
       "      <td>1</td>\n",
       "      <td>chr12</td>\n",
       "      <td>98593756</td>\n",
       "      <td>98593757</td>\n",
       "      <td>392112</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=372600;CLNDISDB=MedGen:CN169374;CLNDN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>392112</td>\n",
       "      <td>392112</td>\n",
       "      <td>AGTGTG</td>\n",
       "      <td>-13</td>\n",
       "      <td>1</td>\n",
       "      <td>chr12</td>\n",
       "      <td>98593756</td>\n",
       "      <td>98593757</td>\n",
       "      <td>392112</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=372600;CLNDISDB=MedGen:CN169374;CLNDN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        0     0_x     1_x  2_x  3_x    0_y       1_y       2_y     3_y  4  5  \\\n",
       "0  228050  228050  AGGTAA   14    1  chr21  44504766  44504767  228050  .  G   \n",
       "1  260298  260298  GGGAGG  -12    1  chr14  77279773  77279774  260298  .  G   \n",
       "2  195517  195517  AGGTAA    8    1  chr17  10533501  10533502  195517  .  A   \n",
       "3  392112  392112  GAGTGT  -14    1  chr12  98593756  98593757  392112  .  C   \n",
       "4  392112  392112  AGTGTG  -13    1  chr12  98593756  98593757  392112  .  C   \n",
       "\n",
       "   6  7                                                  8  \n",
       "0  A  .  AF_EXAC=0.00003;ALLELEID=231377;CLNDISDB=MedGe...  \n",
       "1  A  .  AF_ESP=0.0440;AF_EXAC=0.05999;AF_TGP=0.0184;AL...  \n",
       "2  C  .  AF_ESP=0.0002;AF_EXAC=0.00093;ALLELEID=192678;...  \n",
       "3  A  .  ALLELEID=372600;CLNDISDB=MedGen:CN169374;CLNDN...  \n",
       "4  A  .  ALLELEID=372600;CLNDISDB=MedGen:CN169374;CLNDN...  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We want to merge the Clin Var data together: the info about the mutation and how mutation is affecting ISREs\n",
    "ClinVar_muts_altering_ISREbpProb_DF = pd.DataFrame(ClinVar_muts_altering_ISREbpProb)\n",
    "ClinVar_muts_altering_ISREbpProb_DF[0] = ClinVar_muts_altering_ISREbpProb_DF[0].apply(int)\n",
    "ClinVar_muts_altering_ISREbpProb_DF.head()\n",
    "combined_ClinVar_data = ClinVar_muts_altering_ISREbpProb_DF.merge(clinvar_muts_relevent,left_on=0,right_on=3)\n",
    "combined_ClinVar_data['0_y'] = 'chr' + combined_ClinVar_data['0_y'].astype(str)\n",
    "print combined_ClinVar_data.shape\n",
    "combined_ClinVar_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write ClinVar data to a file\n",
    "combined_ClinVar_data.iloc[:,range(1,14)].to_csv(\"../processed_data/Putative_RiboSPLitches_ClinVarMuts_\"+str(numBasesToFold)+\"bpwindow.txt\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# These are motifs that were tested in Wang et al 2012 papers for ISREs motifs\n",
    "ISE_motifs = ['GGGTTT','GGTGGT','TTTGGG','GAGGGG','GGTATT','GTAACG']\n",
    "ISS_motifs = ['CTCCTC','CTTCAG','GAACAG','ACATGA','AATTTA','AGTAGG','TACAGC','ACAGCT','CAAAGG','AGATAT']\n",
    "ISRE_motifs = ISE_motifs + ISS_motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(247, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2_x</th>\n",
       "      <th>3_x</th>\n",
       "      <th>0_y</th>\n",
       "      <th>1_y</th>\n",
       "      <th>2_y</th>\n",
       "      <th>3_y</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CM095436</td>\n",
       "      <td>CM095436</td>\n",
       "      <td>GGGTTT</td>\n",
       "      <td>-25</td>\n",
       "      <td>3</td>\n",
       "      <td>chr5</td>\n",
       "      <td>70076544</td>\n",
       "      <td>70076545</td>\n",
       "      <td>CM095436</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM;MUT=ALT;GENE=SMN2;STRAND=+;DNA=NM_017...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CM151562</td>\n",
       "      <td>CM151562</td>\n",
       "      <td>TTTGGG</td>\n",
       "      <td>-42</td>\n",
       "      <td>4</td>\n",
       "      <td>chr4</td>\n",
       "      <td>13369887</td>\n",
       "      <td>13369888</td>\n",
       "      <td>CM151562</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM;MUT=ALT;GENE=RAB28;STRAND=-;DNA=NM_00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CM000206</td>\n",
       "      <td>CM000206</td>\n",
       "      <td>CTTCAG</td>\n",
       "      <td>-9</td>\n",
       "      <td>4</td>\n",
       "      <td>chr11</td>\n",
       "      <td>112087915</td>\n",
       "      <td>112087916</td>\n",
       "      <td>CM000206</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM;MUT=ALT;GENE=SDHD;STRAND=+;DNA=NM_003...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CM143109</td>\n",
       "      <td>CM143109</td>\n",
       "      <td>GGTGGT</td>\n",
       "      <td>-38</td>\n",
       "      <td>2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>6006002</td>\n",
       "      <td>6006003</td>\n",
       "      <td>CM143109</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM?;MUT=ALT;GENE=PMS2;STRAND=-;DNA=NM_00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>CM162842</td>\n",
       "      <td>CM162842</td>\n",
       "      <td>AGATAT</td>\n",
       "      <td>-16</td>\n",
       "      <td>2</td>\n",
       "      <td>chr14</td>\n",
       "      <td>87988492</td>\n",
       "      <td>87988493</td>\n",
       "      <td>CM162842</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM;MUT=ALT;GENE=GALC;STRAND=-;DNA=NM_000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0       0_x     1_x  2_x  3_x    0_y        1_y        2_y  \\\n",
       "1   CM095436  CM095436  GGGTTT  -25    3   chr5   70076544   70076545   \n",
       "13  CM151562  CM151562  TTTGGG  -42    4   chr4   13369887   13369888   \n",
       "21  CM000206  CM000206  CTTCAG   -9    4  chr11  112087915  112087916   \n",
       "27  CM143109  CM143109  GGTGGT  -38    2   chr7    6006002    6006003   \n",
       "47  CM162842  CM162842  AGATAT  -16    2  chr14   87988492   87988493   \n",
       "\n",
       "         3_y  4  5  6  7                                                  8  \n",
       "1   CM095436  .  G  C  .  CLASS=DM;MUT=ALT;GENE=SMN2;STRAND=+;DNA=NM_017...  \n",
       "13  CM151562  .  A  C  .  CLASS=DM;MUT=ALT;GENE=RAB28;STRAND=-;DNA=NM_00...  \n",
       "21  CM000206  .  C  T  .  CLASS=DM;MUT=ALT;GENE=SDHD;STRAND=+;DNA=NM_003...  \n",
       "27  CM143109  .  T  C  .  CLASS=DM?;MUT=ALT;GENE=PMS2;STRAND=-;DNA=NM_00...  \n",
       "47  CM162842  .  C  T  .  CLASS=DM;MUT=ALT;GENE=GALC;STRAND=-;DNA=NM_000...  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get mutations that alter ISRE motif bp Probabilities of motifs tested by Wang et al 2012 for HGMD mutations\n",
    "combined_HGMD_data_testedISREs = combined_HGMD_data[combined_HGMD_data['1_x'].isin(ISRE_motifs)]\n",
    "print combined_HGMD_data_testedISREs.shape\n",
    "combined_HGMD_data_testedISREs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort by the absolute value of the column that indicates how far ISRE is from mutation and write a file\n",
    "combined_HGMD_data_testedISREs.reindex(combined_HGMD_data_testedISREs['2_x'].abs().sort_values().index).iloc[:,range(1,14)].to_csv(\"../processed_data/Putative_RiboSPLitches_HGMDMuts_ISREMotifsTested_\"+str(numBasesToFold)+\"bpwindow.txt\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(878, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2_x</th>\n",
       "      <th>3_x</th>\n",
       "      <th>0_y</th>\n",
       "      <th>1_y</th>\n",
       "      <th>2_y</th>\n",
       "      <th>3_y</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>256124</td>\n",
       "      <td>256124</td>\n",
       "      <td>GAGGGG</td>\n",
       "      <td>-18</td>\n",
       "      <td>3</td>\n",
       "      <td>chr19</td>\n",
       "      <td>15187085</td>\n",
       "      <td>15187086</td>\n",
       "      <td>256124</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=256818;CLNDISDB=MedGen:CN169374;CLNDN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>290485</td>\n",
       "      <td>290485</td>\n",
       "      <td>GGGTTT</td>\n",
       "      <td>-9</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>46192508</td>\n",
       "      <td>46192509</td>\n",
       "      <td>290485</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.00003;ALLELEID=274722;CLNDISDB=MedGe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>258492</td>\n",
       "      <td>258492</td>\n",
       "      <td>GAGGGG</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219420459</td>\n",
       "      <td>219420460</td>\n",
       "      <td>258492</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.0378;AF_EXAC=0.03629;AF_TGP=0.0128;AL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>264905</td>\n",
       "      <td>264905</td>\n",
       "      <td>CAAAGG</td>\n",
       "      <td>-22</td>\n",
       "      <td>1</td>\n",
       "      <td>chr13</td>\n",
       "      <td>32321611</td>\n",
       "      <td>32321612</td>\n",
       "      <td>264905</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_TGP=0.0056;ALLELEID=259415;CLNDISDB=MedGen:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>191142</td>\n",
       "      <td>191142</td>\n",
       "      <td>CTCCTC</td>\n",
       "      <td>-24</td>\n",
       "      <td>3</td>\n",
       "      <td>chr2</td>\n",
       "      <td>178721980</td>\n",
       "      <td>178721981</td>\n",
       "      <td>191142</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=188945;CLNDISDB=MedGen:CN517202;CLNDN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         0     0_x     1_x  2_x  3_x    0_y        1_y        2_y     3_y  4  \\\n",
       "33  256124  256124  GAGGGG  -18    3  chr19   15187085   15187086  256124  .   \n",
       "35  290485  290485  GGGTTT   -9    2   chr1   46192508   46192509  290485  .   \n",
       "62  258492  258492  GAGGGG    4    1   chr2  219420459  219420460  258492  .   \n",
       "72  264905  264905  CAAAGG  -22    1  chr13   32321611   32321612  264905  .   \n",
       "83  191142  191142  CTCCTC  -24    3   chr2  178721980  178721981  191142  .   \n",
       "\n",
       "    5  6  7                                                  8  \n",
       "33  C  T  .  ALLELEID=256818;CLNDISDB=MedGen:CN169374;CLNDN...  \n",
       "35  C  G  .  AF_EXAC=0.00003;ALLELEID=274722;CLNDISDB=MedGe...  \n",
       "62  C  A  .  AF_ESP=0.0378;AF_EXAC=0.03629;AF_TGP=0.0128;AL...  \n",
       "72  G  C  .  AF_TGP=0.0056;ALLELEID=259415;CLNDISDB=MedGen:...  \n",
       "83  T  C  .  ALLELEID=188945;CLNDISDB=MedGen:CN517202;CLNDN...  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get mutations that alter ISRE motif bp Probabilities of motifs tested by Wang et al 2012 for ClinVar mutations\n",
    "combined_ClinVar_data_testedISREs = combined_ClinVar_data[combined_ClinVar_data['1_x'].isin(ISRE_motifs)]\n",
    "print combined_ClinVar_data_testedISREs.shape\n",
    "combined_ClinVar_data_testedISREs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort by the absolute value of the column that indicates how far ISRE is from mutation and write a file\n",
    "combined_ClinVar_data_testedISREs.reindex(combined_ClinVar_data_testedISREs['2_x'].abs().sort_values().index).iloc[:,range(1,14)].to_csv(\"../processed_data/Putative_RiboSPLitches_ClinVarMuts_ISREMotifsTested_\"+str(numBasesToFold)+\"bpwindow.txt\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script bash\n",
    "# Let's grab mutations that are not within the introns\n",
    "bedtools intersect -a ../processed_data/CombinedMutations_HGMDandClinVar.bed -b ../processed_data/Introns_DoNotIntersect_Exons.bed -wa > ../temp/CombinedMutations_HGMDandClinVar_Intersect_Introns_DoNotIntersect_Exons.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(124737, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1022187</td>\n",
       "      <td>1022188</td>\n",
       "      <td>263166</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1022187</td>\n",
       "      <td>1022188</td>\n",
       "      <td>263166</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1041949</td>\n",
       "      <td>1041950</td>\n",
       "      <td>128292</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1041949</td>\n",
       "      <td>1041950</td>\n",
       "      <td>128292</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1042189</td>\n",
       "      <td>1042190</td>\n",
       "      <td>263159</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0        1        2       3  4  5       6\n",
       "0  chr1  1022187  1022188  263166  A  G  Benign\n",
       "1  chr1  1022187  1022188  263166  A  G  Benign\n",
       "2  chr1  1041949  1041950  128292  T  C  Benign\n",
       "3  chr1  1041949  1041950  128292  T  C  Benign\n",
       "4  chr1  1042189  1042190  263159  G  A  Benign"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the file that contains the mutations that are intersecting introns that are NOT intersecting any exons\n",
    "muts_Intersect_Introns_DontIntersect_Exons = pd.read_table(\"../temp/CombinedMutations_HGMDandClinVar_Intersect_Introns_DoNotIntersect_Exons.bed\",sep=\"\\t\",header=None)\n",
    "print muts_Intersect_Introns_DontIntersect_Exons.shape\n",
    "muts_Intersect_Introns_DontIntersect_Exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(276, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2_x</th>\n",
       "      <th>3_x</th>\n",
       "      <th>0_y</th>\n",
       "      <th>1_y</th>\n",
       "      <th>2_y</th>\n",
       "      <th>3_y</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CS083955</td>\n",
       "      <td>CS083955</td>\n",
       "      <td>GGTGGG</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>247444845</td>\n",
       "      <td>247444846</td>\n",
       "      <td>CS083955</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM?;MUT=ALT;GENE=NLRP3;STRAND=+;DNA=NM_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CS045477</td>\n",
       "      <td>CS045477</td>\n",
       "      <td>GGTTTG</td>\n",
       "      <td>-13</td>\n",
       "      <td>4</td>\n",
       "      <td>chr22</td>\n",
       "      <td>29636748</td>\n",
       "      <td>29636749</td>\n",
       "      <td>CS045477</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM;MUT=ALT;GENE=NF2;STRAND=+;DNA=NM_0002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>CS064464</td>\n",
       "      <td>CS064464</td>\n",
       "      <td>TTGGTG</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>chr16</td>\n",
       "      <td>31093187</td>\n",
       "      <td>31093188</td>\n",
       "      <td>CS064464</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DP;MUT=ALT;GENE=VKORC1;STRAND=-;DNA=NM_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>CS1510471</td>\n",
       "      <td>CS1510471</td>\n",
       "      <td>AGAGGT</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>chr11</td>\n",
       "      <td>68049425</td>\n",
       "      <td>68049426</td>\n",
       "      <td>CS1510471</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM;MUT=ALT;GENE=TCIRG1;STRAND=+;DNA=NM_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>CS1511525</td>\n",
       "      <td>CS1511525</td>\n",
       "      <td>AGATAG</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154933356</td>\n",
       "      <td>154933357</td>\n",
       "      <td>CS1511525</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM?;MUT=ALT;GENE=F8;STRAND=-;DNA=NM_0001...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            0        0_x     1_x  2_x  3_x    0_y        1_y        2_y  \\\n",
       "8    CS083955   CS083955  GGTGGG    3    1   chr1  247444845  247444846   \n",
       "26   CS045477   CS045477  GGTTTG  -13    4  chr22   29636748   29636749   \n",
       "32   CS064464   CS064464  TTGGTG   42    1  chr16   31093187   31093188   \n",
       "41  CS1510471  CS1510471  AGAGGT    2    3  chr11   68049425   68049426   \n",
       "51  CS1511525  CS1511525  AGATAG   14    1   chrX  154933356  154933357   \n",
       "\n",
       "          3_y  4  5  6  7                                                  8  \n",
       "8    CS083955  .  C  T  .  CLASS=DM?;MUT=ALT;GENE=NLRP3;STRAND=+;DNA=NM_0...  \n",
       "26   CS045477  .  A  C  .  CLASS=DM;MUT=ALT;GENE=NF2;STRAND=+;DNA=NM_0002...  \n",
       "32   CS064464  .  C  G  .  CLASS=DP;MUT=ALT;GENE=VKORC1;STRAND=-;DNA=NM_0...  \n",
       "41  CS1510471  .  T  C  .  CLASS=DM;MUT=ALT;GENE=TCIRG1;STRAND=+;DNA=NM_0...  \n",
       "51  CS1511525  .  A  T  .  CLASS=DM?;MUT=ALT;GENE=F8;STRAND=-;DNA=NM_0001...  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Lets get HGMD muts that are putative RiboSPLitches that do not intersect exons\n",
    "combined_HGMD_data_NoIntersectExons = combined_HGMD_data[combined_HGMD_data[0].isin(muts_Intersect_Introns_DontIntersect_Exons[3])]\n",
    "print combined_HGMD_data_NoIntersectExons.shape\n",
    "combined_HGMD_data_NoIntersectExons.iloc[:,range(1,14)].to_csv(\"../processed_data/Putative_RiboSPLitches_HGMDMuts_DoNotIntersectExons_\"+str(numBasesToFold)+\"bpwindow.txt\",sep=\"\\t\",header=False,index=False)\n",
    "combined_HGMD_data_NoIntersectExons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(29, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2_x</th>\n",
       "      <th>3_x</th>\n",
       "      <th>0_y</th>\n",
       "      <th>1_y</th>\n",
       "      <th>2_y</th>\n",
       "      <th>3_y</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>CS033832</td>\n",
       "      <td>CS033832</td>\n",
       "      <td>GGGTTT</td>\n",
       "      <td>-38</td>\n",
       "      <td>2</td>\n",
       "      <td>chr17</td>\n",
       "      <td>43047634</td>\n",
       "      <td>43047635</td>\n",
       "      <td>CS033832</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM?;MUT=ALT;GENE=BRCA1;STRAND=-;DNA=NM_0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>CR104260</td>\n",
       "      <td>CR104260</td>\n",
       "      <td>CTCCTC</td>\n",
       "      <td>24</td>\n",
       "      <td>4</td>\n",
       "      <td>chr4</td>\n",
       "      <td>15938607</td>\n",
       "      <td>15938608</td>\n",
       "      <td>CR104260</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DFP;MUT=ALT;GENE=FGFBP1;STRAND=-;DB=rs12...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>CR1111332</td>\n",
       "      <td>CR1111332</td>\n",
       "      <td>CTCCTC</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>chr14</td>\n",
       "      <td>53955198</td>\n",
       "      <td>53955199</td>\n",
       "      <td>CR1111332</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DP;MUT=ALT;GENE=BMP4;STRAND=-;DB=rs28555...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>CR1111332</td>\n",
       "      <td>CR1111332</td>\n",
       "      <td>ACAGCT</td>\n",
       "      <td>45</td>\n",
       "      <td>1</td>\n",
       "      <td>chr14</td>\n",
       "      <td>53955198</td>\n",
       "      <td>53955199</td>\n",
       "      <td>CR1111332</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DP;MUT=ALT;GENE=BMP4;STRAND=-;DB=rs28555...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>CS1514361</td>\n",
       "      <td>CS1514361</td>\n",
       "      <td>GAGGGG</td>\n",
       "      <td>-30</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155023026</td>\n",
       "      <td>155023027</td>\n",
       "      <td>CS1514361</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=FP;MUT=ALT;GENE=DCST2;STRAND=-;DNA=NM_14...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0        0_x     1_x  2_x  3_x    0_y        1_y        2_y  \\\n",
       "147   CS033832   CS033832  GGGTTT  -38    2  chr17   43047634   43047635   \n",
       "184   CR104260   CR104260  CTCCTC   24    4   chr4   15938607   15938608   \n",
       "277  CR1111332  CR1111332  CTCCTC    7    1  chr14   53955198   53955199   \n",
       "279  CR1111332  CR1111332  ACAGCT   45    1  chr14   53955198   53955199   \n",
       "286  CS1514361  CS1514361  GAGGGG  -30    2   chr1  155023026  155023027   \n",
       "\n",
       "           3_y  4  5  6  7                                                  8  \n",
       "147   CS033832  .  C  A  .  CLASS=DM?;MUT=ALT;GENE=BRCA1;STRAND=-;DNA=NM_0...  \n",
       "184   CR104260  .  C  A  .  CLASS=DFP;MUT=ALT;GENE=FGFBP1;STRAND=-;DB=rs12...  \n",
       "277  CR1111332  .  G  C  .  CLASS=DP;MUT=ALT;GENE=BMP4;STRAND=-;DB=rs28555...  \n",
       "279  CR1111332  .  G  C  .  CLASS=DP;MUT=ALT;GENE=BMP4;STRAND=-;DB=rs28555...  \n",
       "286  CS1514361  .  G  A  .  CLASS=FP;MUT=ALT;GENE=DCST2;STRAND=-;DNA=NM_14...  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Lets get HGMD muts that are putative RiboSPLitches that are for tested ISREs that do not intersect exons\n",
    "combined_HGMD_data_testedISREs_NoIntersectExons = combined_HGMD_data_testedISREs[combined_HGMD_data_testedISREs[0].isin(muts_Intersect_Introns_DontIntersect_Exons[3])]\n",
    "print combined_HGMD_data_testedISREs_NoIntersectExons.shape\n",
    "combined_HGMD_data_testedISREs_NoIntersectExons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort by the absolute value of the column that indicates how far ISRE is from mutation and write a file\n",
    "combined_HGMD_data_testedISREs_NoIntersectExons.reindex(combined_HGMD_data_testedISREs_NoIntersectExons['2_x'].abs().sort_values().index).iloc[:,range(1,14)].to_csv(\"../processed_data/Putative_RiboSPLitches_HGMDMuts_ISREMotifsTested_DoNotIntersectExons_\"+str(numBasesToFold)+\"bpwindow.txt\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3729, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2_x</th>\n",
       "      <th>3_x</th>\n",
       "      <th>0_y</th>\n",
       "      <th>1_y</th>\n",
       "      <th>2_y</th>\n",
       "      <th>3_y</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>228050</td>\n",
       "      <td>228050</td>\n",
       "      <td>AGGTAA</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>chr21</td>\n",
       "      <td>44504766</td>\n",
       "      <td>44504767</td>\n",
       "      <td>228050</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.00003;ALLELEID=231377;CLNDISDB=MedGe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>260298</td>\n",
       "      <td>260298</td>\n",
       "      <td>GGGAGG</td>\n",
       "      <td>-12</td>\n",
       "      <td>1</td>\n",
       "      <td>chr14</td>\n",
       "      <td>77279773</td>\n",
       "      <td>77279774</td>\n",
       "      <td>260298</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.0440;AF_EXAC=0.05999;AF_TGP=0.0184;AL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>195517</td>\n",
       "      <td>195517</td>\n",
       "      <td>AGGTAA</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10533501</td>\n",
       "      <td>10533502</td>\n",
       "      <td>195517</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.0002;AF_EXAC=0.00093;ALLELEID=192678;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>392117</td>\n",
       "      <td>392117</td>\n",
       "      <td>GGTAAG</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>chr16</td>\n",
       "      <td>23434628</td>\n",
       "      <td>23434629</td>\n",
       "      <td>392117</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=374221;CLNDISDB=MedGen:CN169374;CLNDN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>392117</td>\n",
       "      <td>392117</td>\n",
       "      <td>AGGTAA</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>chr16</td>\n",
       "      <td>23434628</td>\n",
       "      <td>23434629</td>\n",
       "      <td>392117</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=374221;CLNDISDB=MedGen:CN169374;CLNDN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        0     0_x     1_x  2_x  3_x    0_y       1_y       2_y     3_y  4  5  \\\n",
       "0  228050  228050  AGGTAA   14    1  chr21  44504766  44504767  228050  .  G   \n",
       "1  260298  260298  GGGAGG  -12    1  chr14  77279773  77279774  260298  .  G   \n",
       "2  195517  195517  AGGTAA    8    1  chr17  10533501  10533502  195517  .  A   \n",
       "6  392117  392117  GGTAAG    7    3  chr16  23434628  23434629  392117  .  T   \n",
       "7  392117  392117  AGGTAA    8    4  chr16  23434628  23434629  392117  .  T   \n",
       "\n",
       "   6  7                                                  8  \n",
       "0  A  .  AF_EXAC=0.00003;ALLELEID=231377;CLNDISDB=MedGe...  \n",
       "1  A  .  AF_ESP=0.0440;AF_EXAC=0.05999;AF_TGP=0.0184;AL...  \n",
       "2  C  .  AF_ESP=0.0002;AF_EXAC=0.00093;ALLELEID=192678;...  \n",
       "6  G  .  ALLELEID=374221;CLNDISDB=MedGen:CN169374;CLNDN...  \n",
       "7  G  .  ALLELEID=374221;CLNDISDB=MedGen:CN169374;CLNDN...  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Lets get ClinVar muts that are putative RiboSPLitches that do not intersect exons\n",
    "combined_ClinVar_data_NoIntersectExons = combined_ClinVar_data[combined_ClinVar_data[0].apply(str).isin(muts_Intersect_Introns_DontIntersect_Exons[3])]\n",
    "print combined_ClinVar_data_NoIntersectExons.shape\n",
    "combined_ClinVar_data_NoIntersectExons.iloc[:,range(1,14)].to_csv(\"../processed_data/Putative_RiboSPLitches_ClinVarMuts_DoNotIntersectExons_\"+str(numBasesToFold)+\"bpwindow.txt\",sep=\"\\t\",header=False,index=False)\n",
    "combined_ClinVar_data_NoIntersectExons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(337, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2_x</th>\n",
       "      <th>3_x</th>\n",
       "      <th>0_y</th>\n",
       "      <th>1_y</th>\n",
       "      <th>2_y</th>\n",
       "      <th>3_y</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>256124</td>\n",
       "      <td>256124</td>\n",
       "      <td>GAGGGG</td>\n",
       "      <td>-18</td>\n",
       "      <td>3</td>\n",
       "      <td>chr19</td>\n",
       "      <td>15187085</td>\n",
       "      <td>15187086</td>\n",
       "      <td>256124</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=256818;CLNDISDB=MedGen:CN169374;CLNDN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>290485</td>\n",
       "      <td>290485</td>\n",
       "      <td>GGGTTT</td>\n",
       "      <td>-9</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>46192508</td>\n",
       "      <td>46192509</td>\n",
       "      <td>290485</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.00003;ALLELEID=274722;CLNDISDB=MedGe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>258492</td>\n",
       "      <td>258492</td>\n",
       "      <td>GAGGGG</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>chr2</td>\n",
       "      <td>219420459</td>\n",
       "      <td>219420460</td>\n",
       "      <td>258492</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.0378;AF_EXAC=0.03629;AF_TGP=0.0128;AL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>264905</td>\n",
       "      <td>264905</td>\n",
       "      <td>CAAAGG</td>\n",
       "      <td>-22</td>\n",
       "      <td>1</td>\n",
       "      <td>chr13</td>\n",
       "      <td>32321611</td>\n",
       "      <td>32321612</td>\n",
       "      <td>264905</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_TGP=0.0056;ALLELEID=259415;CLNDISDB=MedGen:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>259529</td>\n",
       "      <td>259529</td>\n",
       "      <td>CTCCTC</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>chr22</td>\n",
       "      <td>33316046</td>\n",
       "      <td>33316047</td>\n",
       "      <td>259529</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.0071;AF_EXAC=0.00217;AF_TGP=0.0070;AL...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          0     0_x     1_x  2_x  3_x    0_y        1_y        2_y     3_y  4  \\\n",
       "33   256124  256124  GAGGGG  -18    3  chr19   15187085   15187086  256124  .   \n",
       "35   290485  290485  GGGTTT   -9    2   chr1   46192508   46192509  290485  .   \n",
       "62   258492  258492  GAGGGG    4    1   chr2  219420459  219420460  258492  .   \n",
       "72   264905  264905  CAAAGG  -22    1  chr13   32321611   32321612  264905  .   \n",
       "129  259529  259529  CTCCTC    8    5  chr22   33316046   33316047  259529  .   \n",
       "\n",
       "     5  6  7                                                  8  \n",
       "33   C  T  .  ALLELEID=256818;CLNDISDB=MedGen:CN169374;CLNDN...  \n",
       "35   C  G  .  AF_EXAC=0.00003;ALLELEID=274722;CLNDISDB=MedGe...  \n",
       "62   C  A  .  AF_ESP=0.0378;AF_EXAC=0.03629;AF_TGP=0.0128;AL...  \n",
       "72   G  C  .  AF_TGP=0.0056;ALLELEID=259415;CLNDISDB=MedGen:...  \n",
       "129  G  A  .  AF_ESP=0.0071;AF_EXAC=0.00217;AF_TGP=0.0070;AL...  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Lets get HGMD muts that are putative RiboSPLitches that are for tested ISREs that do not intersect exons\n",
    "combined_ClinVar_data_testedISREs_NoIntersectExons = combined_ClinVar_data_testedISREs[combined_ClinVar_data_testedISREs[0].apply(str).isin(muts_Intersect_Introns_DontIntersect_Exons[3])]\n",
    "print combined_ClinVar_data_testedISREs_NoIntersectExons.shape\n",
    "combined_ClinVar_data_testedISREs_NoIntersectExons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort by the absolute value of the column that indicates how far ISRE is from mutation and write a file\n",
    "combined_ClinVar_data_testedISREs_NoIntersectExons.reindex(combined_ClinVar_data_testedISREs_NoIntersectExons['2_x'].abs().sort_values().index).iloc[:,range(1,14)].to_csv(\"../processed_data/Putative_RiboSPLitches_ClinVarMuts_ISREMotifsTested_DoNotIntersectExons_\"+str(numBasesToFold)+\"bpwindow.txt\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
