{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The purpose of this script is to prepare the mutation datasets to be used for finding RiboSPLitches.\n",
    "\n",
    "The two datasets are from HGMD and ClinVar.\n",
    "\n",
    "The relevant columns we are looking for are chromosome, start and end coordinates, mutID, WT base, MUT base and clinical signficance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import relevant packages\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', 'X', 'Y']\n"
     ]
    }
   ],
   "source": [
    "# Get a list which contains all chromosomes -> valid chromosome list\n",
    "chrom = range(1,23)\n",
    "chrom.extend(['X','Y'])\n",
    "chrom = [str(i) for i in chrom]\n",
    "print chrom"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's start by processing the mutations from the database HGMD. We only want single base mutations which are in the valid chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(169524, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>942142</td>\n",
       "      <td>942143</td>\n",
       "      <td>CM1511864</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM?;MUT=ALT;GENE=SAMD11;STRAND=+;DNA=NM_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>963937</td>\n",
       "      <td>963938</td>\n",
       "      <td>CD142720</td>\n",
       "      <td>.</td>\n",
       "      <td>CCT</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM?;MUT=ALT;GENE=KLHL17;STRAND=+;DNA=NM_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1014142</td>\n",
       "      <td>1014143</td>\n",
       "      <td>CM1411641</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM;MUT=ALT;GENE=ISG15;STRAND=+;DNA=NM_00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1014315</td>\n",
       "      <td>1014316</td>\n",
       "      <td>CI128669</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM;MUT=ALT;GENE=ISG15;STRAND=+;DNA=NM_00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1014358</td>\n",
       "      <td>1014359</td>\n",
       "      <td>CM128668</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>CLASS=DM;MUT=ALT;GENE=ISG15;STRAND=+;DNA=NM_00...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0        1        2          3  4    5   6  7  \\\n",
       "0  1   942142   942143  CM1511864  .    C   G  .   \n",
       "1  1   963937   963938   CD142720  .  CCT   C  .   \n",
       "2  1  1014142  1014143  CM1411641  .    C   T  .   \n",
       "3  1  1014315  1014316   CI128669  .    C  CG  .   \n",
       "4  1  1014358  1014359   CM128668  .    G   T  .   \n",
       "\n",
       "                                                   8  \n",
       "0  CLASS=DM?;MUT=ALT;GENE=SAMD11;STRAND=+;DNA=NM_...  \n",
       "1  CLASS=DM?;MUT=ALT;GENE=KLHL17;STRAND=+;DNA=NM_...  \n",
       "2  CLASS=DM;MUT=ALT;GENE=ISG15;STRAND=+;DNA=NM_00...  \n",
       "3  CLASS=DM;MUT=ALT;GENE=ISG15;STRAND=+;DNA=NM_00...  \n",
       "4  CLASS=DM;MUT=ALT;GENE=ISG15;STRAND=+;DNA=NM_00...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Read all the mutations in HGMD into a dataframe\n",
    "mutations = pd.read_csv(\"../processed_data/HGMD_PRO_2016.3_hg38.bed\",sep=\"\\t\",header=None,low_memory=False)\n",
    "print mutations.shape\n",
    "mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(127445, 9)\n"
     ]
    }
   ],
   "source": [
    "# Get only mutations that are single base mutations [no insertions, deletions]\n",
    "mutations_single = mutations[(mutations[5].isin(['A','G','T','C','N']))&(mutations[6].isin(['A','G','T','C','N']))]\n",
    "print mutations_single.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check that all chromosomes are valid\n",
    "set(mutations_single[0].drop_duplicates().values)==set(chrom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>mutID</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>WTbase</th>\n",
       "      <th>MUTbase</th>\n",
       "      <th>ClinSig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>942142</td>\n",
       "      <td>942143</td>\n",
       "      <td>CM1511864</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>+</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>DM?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1014142</td>\n",
       "      <td>1014143</td>\n",
       "      <td>CM1411641</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>+</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1014358</td>\n",
       "      <td>1014359</td>\n",
       "      <td>CM128668</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>+</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>1022224</td>\n",
       "      <td>1022225</td>\n",
       "      <td>CM148517</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>+</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>1022312</td>\n",
       "      <td>1022313</td>\n",
       "      <td>CM148518</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom    start      end      mutID gene_name strand WTbase MUTbase ClinSig\n",
       "0     1   942142   942143  CM1511864    SAMD11      +      C       G     DM?\n",
       "2     1  1014142  1014143  CM1411641     ISG15      +      C       T      DM\n",
       "4     1  1014358  1014359   CM128668     ISG15      +      G       T      DM\n",
       "5     1  1022224  1022225   CM148517      AGRN      +      G       A      DM\n",
       "6     1  1022312  1022313   CM148518      AGRN      +      A       T      DM"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grab gene name and strand information from column 8 and put into two separate columns\n",
    "# This will be written to a new file \n",
    "mutations_towrite_data = {\"chrom\":mutations_single[0],\"start\":mutations_single[1],\"end\":mutations_single[2],\"mutID\":mutations_single[3],\"gene_name\":mutations_single[8].str.split(pat=';',expand=True)[2].str.split(pat='=',expand=True)[1] ,\"strand\":mutations_single[8].str.split(pat=';',expand=True)[3].str.split(pat='=',expand=True)[1],\"WTbase\":mutations_single[5],\"MUTbase\":mutations_single[6],\"ClinSig\":mutations_single[8].str.split(pat=';',expand=True)[0].str.split(pat='=',expand=True)[1]}\n",
    "#mutations_towrite_data = {\"chrom\":mutations_single[0],\"start\":mutations_single[1],\"end\":mutations_single[2],\"mutID\":mutations_single[3]}\n",
    "mutations_towrite_df = pd.DataFrame(mutations_towrite_data,columns=[\"chrom\",\"start\",\"end\",\"mutID\",\"gene_name\",\"strand\",\"WTbase\",\"MUTbase\",\"ClinSig\"])\n",
    "mutations_towrite_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_towrite_df.to_csv(\"../processed_data/HGMD_PRO_2016_3_hg38_SingleBaseMutations.bed\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next's let's grab the mutations from the database ClinVar. We only want single base mutations which are in the valid chromosomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(346093, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014041</td>\n",
       "      <td>1014042</td>\n",
       "      <td>475283</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=446939;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1014142</td>\n",
       "      <td>1014143</td>\n",
       "      <td>183381</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1014216</td>\n",
       "      <td>1014217</td>\n",
       "      <td>475278</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=446987;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1014227</td>\n",
       "      <td>1014228</td>\n",
       "      <td>402986</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=389314;CLNDISDB=MedGen:CN169374;CLNDN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1014315</td>\n",
       "      <td>1014316</td>\n",
       "      <td>161455</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=171289;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0        1        2       3  4  5   6  7  \\\n",
       "0  1  1014041  1014042  475283  .  G   A  .   \n",
       "1  1  1014142  1014143  183381  .  C   T  .   \n",
       "2  1  1014216  1014217  475278  .  C   T  .   \n",
       "3  1  1014227  1014228  402986  .  G   A  .   \n",
       "4  1  1014315  1014316  161455  .  C  CG  .   \n",
       "\n",
       "                                                   8  \n",
       "0  ALLELEID=446939;CLNDISDB=MedGen:C4015293,OMIM:...  \n",
       "1  ALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:...  \n",
       "2  ALLELEID=446987;CLNDISDB=MedGen:C4015293,OMIM:...  \n",
       "3  ALLELEID=389314;CLNDISDB=MedGen:CN169374;CLNDN...  \n",
       "4  ALLELEID=171289;CLNDISDB=MedGen:C4015293,OMIM:...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read all mutations in ClinVar into a dataframe\n",
    "clinvar_muts = pd.read_csv(\"../processed_data/clinvar.bed\",header=None,sep=\"\\t\",low_memory=False)\n",
    "print clinvar_muts.shape\n",
    "clinvar_muts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(305277, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014041</td>\n",
       "      <td>1014042</td>\n",
       "      <td>475283</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=446939;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1014142</td>\n",
       "      <td>1014143</td>\n",
       "      <td>183381</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1014216</td>\n",
       "      <td>1014217</td>\n",
       "      <td>475278</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=446987;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1014227</td>\n",
       "      <td>1014228</td>\n",
       "      <td>402986</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=389314;CLNDISDB=MedGen:CN169374;CLNDN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>1014358</td>\n",
       "      <td>1014359</td>\n",
       "      <td>161454</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.00001;ALLELEID=171288;CLNDISDB=MedGe...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0        1        2       3  4  5  6  7  \\\n",
       "0  1  1014041  1014042  475283  .  G  A  .   \n",
       "1  1  1014142  1014143  183381  .  C  T  .   \n",
       "2  1  1014216  1014217  475278  .  C  T  .   \n",
       "3  1  1014227  1014228  402986  .  G  A  .   \n",
       "5  1  1014358  1014359  161454  .  G  T  .   \n",
       "\n",
       "                                                   8  \n",
       "0  ALLELEID=446939;CLNDISDB=MedGen:C4015293,OMIM:...  \n",
       "1  ALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:...  \n",
       "2  ALLELEID=446987;CLNDISDB=MedGen:C4015293,OMIM:...  \n",
       "3  ALLELEID=389314;CLNDISDB=MedGen:CN169374;CLNDN...  \n",
       "5  AF_EXAC=0.00001;ALLELEID=171288;CLNDISDB=MedGe...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only get mutations that are single base mutations [so no INDELs]\n",
    "clinvar_muts_single = clinvar_muts[(clinvar_muts[5].isin(['A','G','T','C','N']))&(clinvar_muts[6].isin(['A','G','T','C','N']))]\n",
    "print clinvar_muts_single.shape\n",
    "clinvar_muts_single.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if all rows contain a clinical significant value\n",
    "clinvar_muts_single[clinvar_muts_single[8].str.contains(\"CLNSIG\")].shape == clinvar_muts_single.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We now need to only get mutations that have a clinical significance of pathogenic or likely pathogenic\n",
    "# Need to do some analysis on what kind of clinical significances we observe in this dataset\n",
    "splitUpData = clinvar_muts_single[8].str.split(pat=';',expand=True)\n",
    "storeCLNSIG = []\n",
    "rowsDontContainCLNSIG = []\n",
    "rowsContainCLNSIGINCL = []\n",
    "exceptionRows = []\n",
    "for index,row in splitUpData.iterrows():\n",
    "    try:\n",
    "        y = [i for i in row if i != None and \"CLNSIG=\" in i]\n",
    "        z = [i for i in row if i != None and \"CLNSIG\" in i and \"CLNSIG=\" not in i]\n",
    "        if len(y)==1:\n",
    "            storeCLNSIG.append(row)\n",
    "        elif len(z)==1:\n",
    "            rowsContainCLNSIGINCL.append(row)\n",
    "        else:\n",
    "            rowsDontContainCLNSIG.append(row)\n",
    "    except:\n",
    "        exceptionRows.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "304857\n",
      "420\n",
      "0\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "# Number of rows with no CLNSIG in them\n",
    "print len(storeCLNSIG)\n",
    "print len(rowsContainCLNSIGINCL)\n",
    "print len(rowsDontContainCLNSIG)\n",
    "print len(exceptionRows)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It looks like all entries of ClinVar have a clinical significance associated with them. We just need to extract the clinicial significance from each entry. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function will extract the clinical significance from a given row\n",
    "# Clinical significance can either be in the form \"CLNSIG=\" or \"CLNSIGINCL=\"\n",
    "def getClinicalSigForRow(rowofinterest):\n",
    "    try:\n",
    "        y = [i for i in rowofinterest if i != None and \"CLNSIG=\" in i]\n",
    "        z = [i for i in rowofinterest if i != None and \"CLNSIG\" in i and \"CLNSIG=\" not in i]\n",
    "        if len(y)==1:\n",
    "            return y[0].split('=')[1]\n",
    "        elif len(z)==1:\n",
    "            # Sometimes there are multiple codes associated with CLNSIGINCL, but on cursory glance they appear to be\n",
    "            # the same signficance but with difference codes, so just extracting the first clin sig\n",
    "            return z[0].split(\"=\")[1].split(\"|\")[0].split(':')[1]\n",
    "        else:\n",
    "            print \"Error in getting Clinical Sig for Row in try clause\"\n",
    "    except:\n",
    "        print \"Error in getting Clinical Sig for Row\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split up the extra information column of the clin var data and then extract the clinical significance \n",
    "splitUpData = clinvar_muts_single[8].str.split(pat=';',expand=True)\n",
    "clinSigs = splitUpData.apply(getClinicalSigForRow,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "305277\n"
     ]
    }
   ],
   "source": [
    "# This is the total number of clinSigs extracted \n",
    "print len(clinSigs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>mutID</th>\n",
       "      <th>WTbase</th>\n",
       "      <th>MUTbase</th>\n",
       "      <th>ClinSig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014041</td>\n",
       "      <td>1014042</td>\n",
       "      <td>475283</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1014142</td>\n",
       "      <td>1014143</td>\n",
       "      <td>183381</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1014216</td>\n",
       "      <td>1014217</td>\n",
       "      <td>475278</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1014227</td>\n",
       "      <td>1014228</td>\n",
       "      <td>402986</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>1014358</td>\n",
       "      <td>1014359</td>\n",
       "      <td>161454</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom    start      end   mutID WTbase MUTbase     ClinSig\n",
       "0     1  1014041  1014042  475283      G       A      Benign\n",
       "1     1  1014142  1014143  183381      C       T  Pathogenic\n",
       "2     1  1014216  1014217  475278      C       T      Benign\n",
       "3     1  1014227  1014228  402986      G       A      Benign\n",
       "5     1  1014358  1014359  161454      G       T  Pathogenic"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This will be written to a new file \n",
    "clinvar_muts_towrite_data = {\"chrom\":clinvar_muts_single[0],\"start\":clinvar_muts_single[1],\"end\":clinvar_muts_single[2],\"mutID\":clinvar_muts_single[3],\"WTbase\":clinvar_muts_single[5],\"MUTbase\":clinvar_muts_single[6],\"ClinSig\":clinSigs}\n",
    "#clinvar_muts_towrite_data = {\"chrom\":clinvar_muts_single[0],\"start\":clinvar_muts_single[1],\"end\":clinvar_muts_single[2],\"mutID\":clinvar_muts_single[3]}\n",
    "clinvar_muts_towrite_df = pd.DataFrame(clinvar_muts_towrite_data,columns=[\"chrom\",\"start\",\"end\",\"mutID\",\"WTbase\",\"MUTbase\",\"ClinSig\"])\n",
    "clinvar_muts_towrite_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_muts_towrite_df.to_csv(\"../processed_data/ClinVar_SingleBase_Mutations.bed\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We want to combine the ClinVar data with the HGMD mutation data, but remove any replicates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "!intersectBed -a ../processed_data/HGMD_PRO_2016_3_hg38_SingleBaseMutations.bed -b ../processed_data/ClinVar_SingleBase_Mutations.bed -wa -wb > ../processed_data/HGMD_ClinVar_Intersect.txt\n",
    "!intersectBed -a ../processed_data/HGMD_PRO_2016_3_hg38_SingleBaseMutations.bed -b ../processed_data/ClinVar_SingleBase_Mutations.bed -v > ../processed_data/HGMD_SingleBaseMutations_NotInClinVar.txt\n",
    "!intersectBed -a ../processed_data/ClinVar_SingleBase_Mutations.bed -b ../processed_data/HGMD_PRO_2016_3_hg38_SingleBaseMutations.bed -v > ../processed_data/ClinVar_SingleBaseMutations_NotInHGMD.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(63021, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014142</td>\n",
       "      <td>1014143</td>\n",
       "      <td>CM1411641</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>+</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "      <td>1</td>\n",
       "      <td>1014142</td>\n",
       "      <td>1014143</td>\n",
       "      <td>183381</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1014358</td>\n",
       "      <td>1014359</td>\n",
       "      <td>CM128668</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>+</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "      <td>1</td>\n",
       "      <td>1014358</td>\n",
       "      <td>1014359</td>\n",
       "      <td>161454</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1022224</td>\n",
       "      <td>1022225</td>\n",
       "      <td>CM148517</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>+</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>DM</td>\n",
       "      <td>1</td>\n",
       "      <td>1022224</td>\n",
       "      <td>1022225</td>\n",
       "      <td>243036</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1022312</td>\n",
       "      <td>1022313</td>\n",
       "      <td>CM148518</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "      <td>1</td>\n",
       "      <td>1022312</td>\n",
       "      <td>1022313</td>\n",
       "      <td>243037</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1041581</td>\n",
       "      <td>1041582</td>\n",
       "      <td>CM126385</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>+</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "      <td>1</td>\n",
       "      <td>1041581</td>\n",
       "      <td>1041582</td>\n",
       "      <td>126556</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  0        1        2          3      4  5  6  7   8  9        10       11  \\\n",
       "0  1  1014142  1014143  CM1411641  ISG15  +  C  T  DM  1  1014142  1014143   \n",
       "1  1  1014358  1014359   CM128668  ISG15  +  G  T  DM  1  1014358  1014359   \n",
       "2  1  1022224  1022225   CM148517   AGRN  +  G  A  DM  1  1022224  1022225   \n",
       "3  1  1022312  1022313   CM148518   AGRN  +  A  T  DM  1  1022312  1022313   \n",
       "4  1  1041581  1041582   CM126385   AGRN  +  C  T  DM  1  1041581  1041582   \n",
       "\n",
       "       12 13 14          15  \n",
       "0  183381  C  T  Pathogenic  \n",
       "1  161454  G  T  Pathogenic  \n",
       "2  243036  G  A  Pathogenic  \n",
       "3  243037  A  T  Pathogenic  \n",
       "4  126556  C  T  Pathogenic  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read mutations that are in common between HGMD and ClinVar: they have the same location\n",
    "muts_Both_HGMD_ClinVar = pd.read_csv(\"../processed_data/HGMD_ClinVar_Intersect.txt\",header=None,sep=\"\\t\",low_memory=False)\n",
    "print muts_Both_HGMD_ClinVar.shape\n",
    "muts_Both_HGMD_ClinVar.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(48273, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014142</td>\n",
       "      <td>1014143</td>\n",
       "      <td>CM1411641</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>+</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "      <td>1</td>\n",
       "      <td>1014142</td>\n",
       "      <td>1014143</td>\n",
       "      <td>183381</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1014358</td>\n",
       "      <td>1014359</td>\n",
       "      <td>CM128668</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>+</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "      <td>1</td>\n",
       "      <td>1014358</td>\n",
       "      <td>1014359</td>\n",
       "      <td>161454</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1022224</td>\n",
       "      <td>1022225</td>\n",
       "      <td>CM148517</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>+</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>DM</td>\n",
       "      <td>1</td>\n",
       "      <td>1022224</td>\n",
       "      <td>1022225</td>\n",
       "      <td>243036</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1022312</td>\n",
       "      <td>1022313</td>\n",
       "      <td>CM148518</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "      <td>1</td>\n",
       "      <td>1022312</td>\n",
       "      <td>1022313</td>\n",
       "      <td>243037</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1041581</td>\n",
       "      <td>1041582</td>\n",
       "      <td>CM126385</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>+</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "      <td>1</td>\n",
       "      <td>1041581</td>\n",
       "      <td>1041582</td>\n",
       "      <td>126556</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  0        1        2          3      4  5  6  7   8  9        10       11  \\\n",
       "0  1  1014142  1014143  CM1411641  ISG15  +  C  T  DM  1  1014142  1014143   \n",
       "1  1  1014358  1014359   CM128668  ISG15  +  G  T  DM  1  1014358  1014359   \n",
       "2  1  1022224  1022225   CM148517   AGRN  +  G  A  DM  1  1022224  1022225   \n",
       "3  1  1022312  1022313   CM148518   AGRN  +  A  T  DM  1  1022312  1022313   \n",
       "4  1  1041581  1041582   CM126385   AGRN  +  C  T  DM  1  1041581  1041582   \n",
       "\n",
       "       12 13 14          15  \n",
       "0  183381  C  T  Pathogenic  \n",
       "1  161454  G  T  Pathogenic  \n",
       "2  243036  G  A  Pathogenic  \n",
       "3  243037  A  T  Pathogenic  \n",
       "4  126556  C  T  Pathogenic  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grab the mutations that have the same base changes in both HGMD and ClinVar\n",
    "muts_Both_HGMD_ClinVar_SameBase = muts_Both_HGMD_ClinVar[(muts_Both_HGMD_ClinVar[6]==muts_Both_HGMD_ClinVar[13])&(muts_Both_HGMD_ClinVar[7]==muts_Both_HGMD_ClinVar[14])]\n",
    "print muts_Both_HGMD_ClinVar_SameBase.shape\n",
    "muts_Both_HGMD_ClinVar_SameBase.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14748, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1</td>\n",
       "      <td>1806502</td>\n",
       "      <td>1806503</td>\n",
       "      <td>CM164924</td>\n",
       "      <td>GNB1</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>DM</td>\n",
       "      <td>1</td>\n",
       "      <td>1806502</td>\n",
       "      <td>1806503</td>\n",
       "      <td>391609</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1</td>\n",
       "      <td>1806502</td>\n",
       "      <td>1806503</td>\n",
       "      <td>CM164924</td>\n",
       "      <td>GNB1</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>DM</td>\n",
       "      <td>1</td>\n",
       "      <td>1806502</td>\n",
       "      <td>1806503</td>\n",
       "      <td>224715</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1</td>\n",
       "      <td>1806502</td>\n",
       "      <td>1806503</td>\n",
       "      <td>CM164922</td>\n",
       "      <td>GNB1</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "      <td>1</td>\n",
       "      <td>1806502</td>\n",
       "      <td>1806503</td>\n",
       "      <td>391609</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1</td>\n",
       "      <td>1806502</td>\n",
       "      <td>1806503</td>\n",
       "      <td>CM164922</td>\n",
       "      <td>GNB1</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "      <td>1</td>\n",
       "      <td>1806502</td>\n",
       "      <td>1806503</td>\n",
       "      <td>208722</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Conflicting_interpretations_of_pathogenicity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1</td>\n",
       "      <td>2228824</td>\n",
       "      <td>2228825</td>\n",
       "      <td>CM165733</td>\n",
       "      <td>SKI</td>\n",
       "      <td>+</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>DM</td>\n",
       "      <td>1</td>\n",
       "      <td>2228824</td>\n",
       "      <td>2228825</td>\n",
       "      <td>409969</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0        1        2         3     4  5  6  7   8  9        10       11  \\\n",
       "28  1  1806502  1806503  CM164924  GNB1  -  A  G  DM  1  1806502  1806503   \n",
       "30  1  1806502  1806503  CM164924  GNB1  -  A  G  DM  1  1806502  1806503   \n",
       "31  1  1806502  1806503  CM164922  GNB1  -  A  T  DM  1  1806502  1806503   \n",
       "32  1  1806502  1806503  CM164922  GNB1  -  A  T  DM  1  1806502  1806503   \n",
       "40  1  2228824  2228825  CM165733   SKI  +  C  A  DM  1  2228824  2228825   \n",
       "\n",
       "        12 13 14                                            15  \n",
       "28  391609  A  C                                    Pathogenic  \n",
       "30  224715  A  T                                    Pathogenic  \n",
       "31  391609  A  C                                    Pathogenic  \n",
       "32  208722  A  G  Conflicting_interpretations_of_pathogenicity  \n",
       "40  409969  C  T                        Uncertain_significance  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grab the mutations that have do not have the same base changes in both HGMD and ClinVar\n",
    "muts_Both_HGMD_ClinVar_DiffBase_SameLoc = muts_Both_HGMD_ClinVar[(muts_Both_HGMD_ClinVar[6]!=muts_Both_HGMD_ClinVar[13])|(muts_Both_HGMD_ClinVar[7]!=muts_Both_HGMD_ClinVar[14])]\n",
    "print muts_Both_HGMD_ClinVar_DiffBase_SameLoc.shape\n",
    "muts_Both_HGMD_ClinVar_DiffBase_SameLoc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5561\n",
      "5787\n"
     ]
    }
   ],
   "source": [
    "# The problem above as we can notice is that using bedtools intersectBind, it's overlapping the same line in file\n",
    "# with multiple lines in the other file and vice versa. \n",
    "# Quick way to check is if IDs found in dataset: same location, different base has any IDs in dataset: same base,\n",
    "# same location\n",
    "# Those that don't will be included as unique rows\n",
    "#ClinVar_IDs_ThatDiffer = [i for i in list(muts_Both_HGMD_ClinVar_DiffBase_SameLoc[12].values) if i not in list(muts_Both_HGMD_ClinVar_SameBase[12].values)]\n",
    "ClinVar_IDs_ThatDiffer = list(muts_Both_HGMD_ClinVar_DiffBase_SameLoc[~muts_Both_HGMD_ClinVar_DiffBase_SameLoc[12].isin(muts_Both_HGMD_ClinVar_SameBase[12])][12].values)\n",
    "print len(ClinVar_IDs_ThatDiffer)\n",
    "HGMD_IDs_ThatDiffer = ClinVar_IDs_ThatDiffer = list(muts_Both_HGMD_ClinVar_DiffBase_SameLoc[~muts_Both_HGMD_ClinVar_DiffBase_SameLoc[3].isin(muts_Both_HGMD_ClinVar_SameBase[3])][3].values)\n",
    "print len(HGMD_IDs_ThatDiffer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gather all mutations in both HGMD and ClinVar that have the same coordinates with similar named columns\n",
    "muts_Both_SameBase = muts_Both_HGMD_ClinVar_SameBase.iloc[:,[0,1,2,3,6,7,8]]\n",
    "muts_Both_SameBase.columns = [\"Chrom\",\"Start\",\"End\",\"MutID\",\"WTbase\",\"MUTbase\",\"ClinSig\"]\n",
    "muts_Both_DiffBase_HGMD = muts_Both_HGMD_ClinVar_DiffBase_SameLoc[muts_Both_HGMD_ClinVar_DiffBase_SameLoc[3].isin(HGMD_IDs_ThatDiffer)].iloc[:,[0,1,2,3,6,7,8]]\n",
    "muts_Both_DiffBase_HGMD.columns = [\"Chrom\",\"Start\",\"End\",\"MutID\",\"WTbase\",\"MUTbase\",\"ClinSig\"]\n",
    "muts_Both_DiffBase_ClinVar = muts_Both_HGMD_ClinVar_DiffBase_SameLoc[muts_Both_HGMD_ClinVar_DiffBase_SameLoc[12].isin(ClinVar_IDs_ThatDiffer)].iloc[:,[9,10,11,12,13,14,15]]\n",
    "muts_Both_DiffBase_ClinVar.columns = [\"Chrom\",\"Start\",\"End\",\"MutID\",\"WTbase\",\"MUTbase\",\"ClinSig\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(73798, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>942142</td>\n",
       "      <td>942143</td>\n",
       "      <td>CM1511864</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>+</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>DM?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1211916</td>\n",
       "      <td>1211917</td>\n",
       "      <td>CS060109</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>DP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1232512</td>\n",
       "      <td>1232513</td>\n",
       "      <td>CM1411602</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1232753</td>\n",
       "      <td>1232754</td>\n",
       "      <td>CM148989</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>+</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1232833</td>\n",
       "      <td>1232834</td>\n",
       "      <td>CM165283</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>+</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0        1        2          3        4  5  6  7    8\n",
       "0  1   942142   942143  CM1511864   SAMD11  +  C  G  DM?\n",
       "1  1  1211916  1211917   CS060109  TNFRSF4  -  G  A   DP\n",
       "2  1  1232512  1232513  CM1411602  B3GALT6  +  A  G   DM\n",
       "3  1  1232753  1232754   CM148989  B3GALT6  +  C  A   DM\n",
       "4  1  1232833  1232834   CM165283  B3GALT6  +  T  C   DM"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read mutations that are only found in HGMD\n",
    "muts_Only_HGMD = pd.read_csv(\"../processed_data/HGMD_SingleBaseMutations_NotInClinVar.txt\",header=None,sep=\"\\t\",low_memory=False)\n",
    "print muts_Only_HGMD.shape\n",
    "muts_Only_HGMD.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(252132, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014041</td>\n",
       "      <td>1014042</td>\n",
       "      <td>475283</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1014216</td>\n",
       "      <td>1014217</td>\n",
       "      <td>475278</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1014227</td>\n",
       "      <td>1014228</td>\n",
       "      <td>402986</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1014450</td>\n",
       "      <td>1014451</td>\n",
       "      <td>475281</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1014470</td>\n",
       "      <td>1014471</td>\n",
       "      <td>475282</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Likely_benign</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0        1        2       3  4  5              6\n",
       "0  1  1014041  1014042  475283  G  A         Benign\n",
       "1  1  1014216  1014217  475278  C  T         Benign\n",
       "2  1  1014227  1014228  402986  G  A         Benign\n",
       "3  1  1014450  1014451  475281  C  T         Benign\n",
       "4  1  1014470  1014471  475282  G  C  Likely_benign"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read mutations that are only found in ClinVar\n",
    "muts_Only_ClinVar = pd.read_csv(\"../processed_data/ClinVar_SingleBaseMutations_NotInHGMD.txt\",header=None,sep=\"\\t\",low_memory=False)\n",
    "print muts_Only_ClinVar.shape\n",
    "muts_Only_ClinVar.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gather all mutations that are only in HGMD or ClinVar with similar named columns\n",
    "muts_Only_HGMD_NoStrand = muts_Only_HGMD.iloc[:,[0,1,2,3,6,7,8]]\n",
    "muts_Only_HGMD_NoStrand.columns = [\"Chrom\",\"Start\",\"End\",\"MutID\",\"WTbase\",\"MUTbase\",\"ClinSig\"]\n",
    "muts_Only_ClinVar.columns = [\"Chrom\",\"Start\",\"End\",\"MutID\",\"WTbase\",\"MUTbase\",\"ClinSig\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(379990, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>MutID</th>\n",
       "      <th>WTbase</th>\n",
       "      <th>MUTbase</th>\n",
       "      <th>ClinSig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014142</td>\n",
       "      <td>1014143</td>\n",
       "      <td>CM1411641</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1014358</td>\n",
       "      <td>1014359</td>\n",
       "      <td>CM128668</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1022224</td>\n",
       "      <td>1022225</td>\n",
       "      <td>CM148517</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1022312</td>\n",
       "      <td>1022313</td>\n",
       "      <td>CM148518</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1041581</td>\n",
       "      <td>1041582</td>\n",
       "      <td>CM126385</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chrom    Start      End      MutID WTbase MUTbase ClinSig\n",
       "0     1  1014142  1014143  CM1411641      C       T      DM\n",
       "1     1  1014358  1014359   CM128668      G       T      DM\n",
       "2     1  1022224  1022225   CM148517      G       A      DM\n",
       "3     1  1022312  1022313   CM148518      A       T      DM\n",
       "4     1  1041581  1041582   CM126385      C       T      DM"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Append all the data frames together to get a total set of mutations to work with \n",
    "allmuts = muts_Both_SameBase.append(muts_Both_DiffBase_HGMD.append(muts_Both_DiffBase_ClinVar.append(muts_Only_HGMD_NoStrand.append(muts_Only_ClinVar))))\n",
    "print allmuts.shape\n",
    "allmuts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(379990, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>MutID</th>\n",
       "      <th>WTbase</th>\n",
       "      <th>MUTbase</th>\n",
       "      <th>ClinSig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>942142</td>\n",
       "      <td>942143</td>\n",
       "      <td>CM1511864</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>DM?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014041</td>\n",
       "      <td>1014042</td>\n",
       "      <td>475283</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014142</td>\n",
       "      <td>1014143</td>\n",
       "      <td>CM1411641</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014216</td>\n",
       "      <td>1014217</td>\n",
       "      <td>475278</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014227</td>\n",
       "      <td>1014228</td>\n",
       "      <td>402986</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chrom    Start      End      MutID WTbase MUTbase ClinSig\n",
       "0  chr1   942142   942143  CM1511864      C       G     DM?\n",
       "0  chr1  1014041  1014042     475283      G       A  Benign\n",
       "0  chr1  1014142  1014143  CM1411641      C       T      DM\n",
       "1  chr1  1014216  1014217     475278      C       T  Benign\n",
       "2  chr1  1014227  1014228     402986      G       A  Benign"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sort the data and add \"chr\" to the chromosome name\n",
    "allmuts_sorted = allmuts.sort_values([\"Chrom\",\"Start\"])\n",
    "allmuts_sorted['Chrom'] = 'chr' + allmuts_sorted['Chrom'].astype(str)\n",
    "print allmuts_sorted.shape\n",
    "allmuts_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the data to a new file\n",
    "allmuts_sorted.to_csv(\"../processed_data/CombinedMutations_HGMDandClinVar.bed\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's separate out HGMD and ClinVar information into two different files. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(127858, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>MutID</th>\n",
       "      <th>Strand</th>\n",
       "      <th>WTbase</th>\n",
       "      <th>MUTbase</th>\n",
       "      <th>ClinSig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>942142</td>\n",
       "      <td>942143</td>\n",
       "      <td>CM1511864</td>\n",
       "      <td>+</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>DM?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014142</td>\n",
       "      <td>1014143</td>\n",
       "      <td>CM1411641</td>\n",
       "      <td>+</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014358</td>\n",
       "      <td>1014359</td>\n",
       "      <td>CM128668</td>\n",
       "      <td>+</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1022224</td>\n",
       "      <td>1022225</td>\n",
       "      <td>CM148517</td>\n",
       "      <td>+</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1022312</td>\n",
       "      <td>1022313</td>\n",
       "      <td>CM148518</td>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>DM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chrom    Start      End      MutID Strand WTbase MUTbase ClinSig\n",
       "0  chr1   942142   942143  CM1511864      +      C       G     DM?\n",
       "0  chr1  1014142  1014143  CM1411641      +      C       T      DM\n",
       "1  chr1  1014358  1014359   CM128668      +      G       T      DM\n",
       "2  chr1  1022224  1022225   CM148517      +      G       A      DM\n",
       "3  chr1  1022312  1022313   CM148518      +      A       T      DM"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This is for HGMD mut\n",
    "muts_Both_SameBase = muts_Both_HGMD_ClinVar_SameBase.iloc[:,[0,1,2,3,5,6,7,8]]\n",
    "muts_Both_SameBase.columns = [\"Chrom\",\"Start\",\"End\",\"MutID\",\"Strand\",\"WTbase\",\"MUTbase\",\"ClinSig\"]\n",
    "muts_Both_DiffBase_HGMD = muts_Both_HGMD_ClinVar_DiffBase_SameLoc[muts_Both_HGMD_ClinVar_DiffBase_SameLoc[3].isin(HGMD_IDs_ThatDiffer)].iloc[:,[0,1,2,3,5,6,7,8]]\n",
    "muts_Both_DiffBase_HGMD.columns = [\"Chrom\",\"Start\",\"End\",\"MutID\",\"Strand\",\"WTbase\",\"MUTbase\",\"ClinSig\"]\n",
    "muts_Only_HGMD_Strand = muts_Only_HGMD.iloc[:,[0,1,2,3,5,6,7,8]]\n",
    "muts_Only_HGMD_Strand.columns = [\"Chrom\",\"Start\",\"End\",\"MutID\",\"Strand\",\"WTbase\",\"MUTbase\",\"ClinSig\"]\n",
    "allmuts_HGMD = muts_Both_SameBase.append(muts_Only_HGMD_Strand.append(muts_Both_DiffBase_HGMD))\n",
    "allmuts_HGMD_sorted = allmuts_HGMD.sort_values([\"Chrom\",\"Start\"])\n",
    "allmuts_HGMD_sorted['Chrom'] = 'chr' + allmuts_HGMD_sorted['Chrom'].astype(str)\n",
    "print allmuts_HGMD_sorted.shape\n",
    "allmuts_HGMD_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the data to a new file\n",
    "allmuts_HGMD_sorted.to_csv(\"../processed_data/Mutations_HGMD.bed\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(252132, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>MutID</th>\n",
       "      <th>WTbase</th>\n",
       "      <th>MUTbase</th>\n",
       "      <th>ClinSig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014041</td>\n",
       "      <td>1014042</td>\n",
       "      <td>475283</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014216</td>\n",
       "      <td>1014217</td>\n",
       "      <td>475278</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014227</td>\n",
       "      <td>1014228</td>\n",
       "      <td>402986</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014450</td>\n",
       "      <td>1014451</td>\n",
       "      <td>475281</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1014470</td>\n",
       "      <td>1014471</td>\n",
       "      <td>475282</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Likely_benign</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chrom    Start      End   MutID WTbase MUTbase        ClinSig\n",
       "0  chr1  1014041  1014042  475283      G       A         Benign\n",
       "1  chr1  1014216  1014217  475278      C       T         Benign\n",
       "2  chr1  1014227  1014228  402986      G       A         Benign\n",
       "3  chr1  1014450  1014451  475281      C       T         Benign\n",
       "4  chr1  1014470  1014471  475282      G       C  Likely_benign"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This is the ClinVar data\n",
    "allmuts_ClinVar = muts_Only_ClinVar.append(muts_Both_DiffBase_ClinVar)\n",
    "allmuts_ClinVar_sorted = allmuts_ClinVar.sort_values([\"Chrom\",\"Start\"])\n",
    "allmuts_ClinVar_sorted['Chrom'] = 'chr' + allmuts_ClinVar_sorted['Chrom'].astype(str)\n",
    "print allmuts_ClinVar_sorted.shape\n",
    "allmuts_ClinVar_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the data to a new file\n",
    "allmuts_ClinVar_sorted.to_csv(\"../processed_data/Mutations_ClinVar.bed\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
